---
title: "Microbiome_vignette"
output: 
  rmarkdown::html_vignette:
    fig_width: 7
    fig_height: 7
    fig_align: center
    text_align: justify
    author: "RÃ©my Villette"
    toc: true
    number_section: true
    warning: false
    
vignette: >
  %\VignetteIndexEntry{Microbiome_vignette}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

```

```{r setup, message=F, warning=F}
library(Microbiome)
library(phyloseq)
```

# Introduction
<div style="text-align: justify">

This package is based on the code of Dr. Villette and Dr. Larsen. The package is written and maintained by Dr. Villette. This package is meant to facilitate microbiome exploration and ensuring nice plotting.\
This package covers :

-   The dada2 pipeline with wrapper functions that ease the processing of multiple projects

-   Some plotting functions for beta diversity, heatmap and differential abundance analysis using directly a phyloseq object

-   A pipeline for IgASeq analysis

</div>

# Trim, denoise and align your sequences using dada2 pipeline
<div style="text-align: justify">
For convenience this will not be a reproducible example, dada2 takes too long to compute and knit. This part of the tutorial will present a run that we performed in house. The rest of the tutorial will be based on reproducible data.
</div>

## Get the files
<div style="text-align: justify">
We use here a wrapper function that will create a list of three for pair end :

-   forward files

-   reverse files

-   names of the files

And for single end : -

-   a list of files

-   names of the files
</div>
```{r eval=F}

f_list = list_fastq("/home/bigbeast/Documents/tmp/2022-12 CIMMAP run ELISE", 
                    pattern = c("R1", "R2"), separator = "_", level = 1)

# check that all files are distinct

lapply(f_list, duplicated)

# check that all files exist5
lapply(f_list, file.exists)
random=  lapply(f_list, "[",sample(1:255,30))
# tmp= summarise_fastq(random, cores =30, plot = F )
```

## Check the quality profile
<div style="text-align: justify">
We will use the function `qc_check`. This will take time as the `plotQualityProfile` isn't parallelized in dada2. This function will create two plots (for pair end) of `n` aggregated samples and only one plot if you are using single end.
</div>
```{r eval=F}
qc_check(flist, n=30)
```

## Check the best trimming parameters for your set of fastq

### Test optimal cutting param

```{r include=T, eval=F}
set.seed(1)
tmp = lapply(f_list, "[",sample(1:255,30))
tmp_list= filt_list(tmp)
filtered= list()

cb= combn(x= (300-seq(0, 50, by=10)), m=2)
# Very long be careful
#try to test combination of trimmings
for(i in 1:dim(cb)[2]){
  filtered[[i]]= filter_fastq(tmp, tmp_list, cutting_param = cb[ ,i], cores = 35, trimleft = 35, maxEE = c(3,4))
}
#extract the data to a df 
per= NULL
for(i in 1:15) {
t =  filtered[[i]] %>% 
  as.data.frame() %>%
  mutate(per=reads.out/reads.in)
per= cbind(per, t$per) %>% 
  as.data.frame()
# per$names=rownames(tmp[1])
print(per)
}

colnames(per)= paste(cb[1,], cb[2,]) 
per$names=rownames(t[1])

# plot 

png("transmic and IgA rescue mice percent reads passed depending on cutting parameters.png", width = 600, height=400)
per %>% 
  as.data.frame() %>%
  pivot_longer(names_to = "cut", values_to = "per", cols = 1:15) %>% 
  ggplot(aes(y=per, x= cut))+
  geom_boxplot()+ geom_line(aes(group=names, col=names))+
  labs(y= "percentage passed", x= "Trimming parameters")+
  theme(axis.text.x = element_text(angle=90), legend.position = "none")
 

dev.off()
```

### Test optimal maxEE

```{r, eval=F, include=T}
cb= combn(x= (7-seq(0,5, by=1)), m=2)
cb =  cb[nrow(cb):1,]
# Very long be careful
#try to test combination of trimmings
# registerDoParallel(cl = makeCluster(5))
for(i in 1:dim(cb)[2]){
  filtered[[i]]= filter_fastq(tmp, tmp_list, cutting_param = c(260,250),
                              cores = 35, trimleft = 35, maxEE = cb[ ,i])
}
#extract the data to a df 
per= NULL
for(i in 1:15) {
t =  filtered[[i]] %>% 
  as.data.frame() %>%
  mutate(per=reads.out/reads.in)
per= cbind(per, t$per) %>% 
  as.data.frame()
}

colnames(per)= paste(cb[1,], cb[2,]) 
per$names=rownames(t[1])

# plot 

png("transmic and IgA rescue mice percent reads passed depending on errors parameters.png", width = 600, height=400)
per %>% as.data.frame() %>% pivot_longer(names_to = "cut", values_to = "per", cols = 1:15) %>% 
  ggplot(aes(y=per, x= cut))+
  geom_boxplot()+ geom_line(aes(group=names, col=names))+
  labs(y= "percentage passed", x= "Trimming parameters")+
  theme(axis.text.x = element_text(angle=90), legend.position = "none")
dev.off()


```

## Filter and trim the fastq
<div style="text-align: justify">
We now have to remove the bad quality reads and trim the length. You will find a function to create the list of filtered files and one to make the filtered files.
</div>
```{r eval=F}
filt= filt_list(f_list) # create the list of filtered files

filtered = filterAndTrim(fwd= fwd, filt = filtFs, rev=rv, filt.rev = filtRs, truncLen = c(260,240), trimLeft = 25, maxEE = c(3,5), multithread = 45)
```

# Analyse your 16S data now
<div style="text-align: justify">
We will use the **enterotype** data to explore some of the plotting functions. Let's start with the beta diversity functions `beta_diversity` and `beta_dispersion`.
</div>
```{r}
data(enterotype)

```

## Alpha diversity plots

TBD

## Beta diversity plots
<div style="text-align: justify">
You will have the choice between `beta_diversity` and `beta_dispersion` for your beta diversity plotting.

\code{beta_diversity} will plot any of the following methods : PCoA, NMDS, PCA, BCA. More methods will be implemented later on. You will end up with a two dimensions projected using dots and ellipses. **The ellipses here are not confidence ellipses but a graphical summary**.

\code{beta_dispersion} will plot only PCoA for the moment, I need to work around some errors to implement more methods. This function will plot the two components of your choosing, **confidence ellipses** and boxplot for each axis and for each group. Each function will return a plot and a percentage of contribution for each component.
</div>
```{r message=F, warning=F}
beta_diversity(enterotype, dist="bray", method="NMDS", group="SeqTech", permanova = F)

beta_dispersion(enterotype, dist = "bray", method = "PCoA", group = "SeqTech")
```

You can play with the parameters of each function like the following plots :

```{r message=F, warning=F}
beta_diversity(enterotype, dist="bray", method="PCoA", group="SeqTech",
               color_vector =c("#777711", "#117777", "#DD7788"),
               factor_to_plot = "Sequencing technologies \n impact 16S results", 
               lwd = 2, cpoint = 2, permanova = T)

beta_dispersion(enterotype, dist = "bray", method = "PCoA", group = "SeqTech",
                color_vector = c("#777711", "#117777", "#DD7788"),
                legend_title = "Sequencing tech", lwd = 2, 
                font = 2, draw = "polygon", text = T, permanova = T, 
                y.intersp = 0.7)

```

## Heatmap based on the phyloseq object
<div style="text-align: justify">
This function will perform a top taxa at the rank you choose and create a heatmap with annotations. For now only one annotation is supported.\
The clusterisation is made using the `hclust`function and a Ward.D2 method. You can define the distance matrix that you want to use. It is important to note that the distance is made before any trimming of the data. This means that the distance matrix is made at the ASV/OTU level before doing the rank merging and topping, so the clusterization will represent your "true" data instead of a modified dataset.

```{r message=F, fig.width=15, fig.height=10}
phylo_heatmap(enterotype, top = 30, labels = "SeqTech", taxa_rank = "Genus", factor_to_plot ="Enterotype", split = 3, distance = "bray" )
```

The idea behind these functions is : creating a more automatic pipeline enabling filtering and subseting of the phyloseq object without having to perform a temporary phyloseq object and a temporary distance object. With these function you can directly use the "pipe" introduce by `magrittr` and `dplyr.`\
So you can use a `subset_samples` like the following and automatically plot your beta diversity without adding to much code.

```{r message=F, warning=F}
enterotype %>%
  subset_samples(SeqTech=="Illumina") %>% 
  beta_dispersion(group="Enterotype", color_vector = c("#777711", "#117777", "#DD7788"), 
                  legend_title = "Enterotypes \n with bray" ,
                  lwd = 2, font = 2, draw = "lines", text = T, permanova = T, y.intersp = 0.7, 
                  where="bottomleft", cex = 3)
```

Additionnaly, if you want to go further you can also serialized the code with a `for` loop or a `lapply` or even a parallel approach using `mclapply` or `doParallel`.

```{r message= F, warning=F, fig.height=5}
layout(matrix(c(1,2,3),
                nrow = 1,
                ncol = 3,
                byrow = TRUE))
for(i in c("Illumina", "Sanger", "Pyro454")){
  enterotype %>%
    subset_samples(SeqTech==i & !is.na(Enterotype))%>%
    beta_diversity(dist="bray", method="PCoA", group="Enterotype", 
                   color_vector =c("#777711", "#117777", "#DD7788"), 
                   factor_to_plot = paste("Enterotypes using \n",i, "sequencing"), lwd = 2, cpoint = 2)
}

```
</div>
## Differential abundance testing {.tabset}

```{r message=F}
data("GlobalPatterns")
tmp= GlobalPatterns %>%
  subset_samples(SampleType=="Feces" | SampleType=="Soil")%>%
  tax_glom("Genus")
res= tmp %>% 
  differential_abundance( group="SampleType", col1 = "brown", col2="darkgreen")

```
### Results
```{r}
head(res$all_features)
```
### Barplot results
```{r}
res$barplot
```

### Volcano plot results
```{r}
res$volcano
```

TBD

## Model testing
<div style="text-align: justify">
Last but not least : machine learning. Machine learning is used in general to predict outcome or predict class assignation between health and disease status. For this purpose we developed wrapper functions to screen models more easily. Most of the code is actually generated using **caret** [https://topepo.github.io/caret/] package. \

For now the functions implement *randomForest*, *glmnet* and *plsda* models, other model might work but I didn't test them yet. 

### Screen models on your raw data
<div style="text-align: justify">
This function accepts either <phyloseq> objects or a list containing a matrix of taxa for example and a dataframe with the variables to test. 
```{r}
res = screen_models(enterotype, model="glmnet", cores = 1, number = 3, repeats = 3)
```


</div>
# IgA seq analysis pipeline
<div style="text-align: justify">
You've performed all your sortings and sequencing, you now have three samples coming from a single individual.\
We created a pipeline analysis where you will use the **neg1** (9/10 of the neg fraction) and **neg2** (1/10 of the neg fraction) dispersion (centered and reduced) to create a normal dispersion, using a Z approach we will have a Z score for the `pos/neg1` based on the standard deviation of `neg1/neg2`.

Here we can see what the analysis will look like:


```{r echo=F, fig.width=7, fig.height=7}
knitr::include_graphics("/home/remy/Images/Z test.png")
```

-   The technical dispersion is assessed using `neg1/neg2`for the log2 ratio and the `neg1*neg2`abundance for log10 (black dot)

-   The biological dispersion is assessed `pos/neg1` for the log2 ratio and the `pos*neg1` abundance for log10 (orange dot)

We will then take windows of X ASV (for example 20) to create n Gaussian curve and n ...\
The pipeline is based on three main functions that will call for other functions : seq_table, slide_z and collapse_IgAseq.

</div>
## Make the seq_table list
<div style="text-align: justify">
First the seq_table function will take your phyloseq object and transform it to a list of data frames, one data frame for each samples coming from a single individual. For this function you will need to give physeq, sample_name corresponding to the name identifying the individual from which the sorted samples came, sorting_names the column where we find the samples such as : "sample1_pos", "sample1_neg1", "sample1_neg2". Then the cols_to_keep that need to stay for now on "all", it will collapse your sample_data in one column to allow you get it back later on.

The function will tell you if there is some samples are alone, if you have duplicated samples you need to sort them out or the rest of the pipeline will block. You can take a look at the architecture of the new object, you will find the ASV sequence as rownames, the taxonomy collapsed with "\#" separator, the three samples having their own column and the sample_data collapse using also "\#" separator. The function will only take the ASV that are present in the samples.

```{r eval=T}
 data("igaseq")
sample_names(igaseq)= sample_data(igaseq)$sample_sort

seq.tab= seq_table(igaseq, sample_name = "sample_origin", sorting_names = "sample_sort", cols_to_keep = "all" )
```

```{r eval=F}

DT::datatable(seq$MO101, rownames = F)

```

The colnames will have the sample_names as given in the otu_table(), check that they correspond to your pos, neg1 and neg2.

Run the main function Now we will run the main function : slide_z. This function will take you seq_table object and run the Z function for each samples. If you are running this function for the first time use plot=T, if you already made the plots let it as FALSE it will make the loop a lot faster.

</div>

## Deal with the zero values
<div style="text-align: justify">
In this approach we will use log2 ratio and log10 of abundances. As you know log2(0/x) or log2(x/0) can't be performed, so we need a way to deal with the zeros. We came up with two approach :

-   remove all ASV with a zero value

-   replace 0 by a random number between 0 and the min value found in one of the three samples

-   replace 0 by the minimum count in each samples, this probably the worst thing to do

In sample with few ASV, i.e meconiums, I strongly recommend using the random_generation while in adults samples the no_zero approach is performing better. For more complex samples the decision is up to you, you can use the function neg_dipsersion to visualize the two outcomes.

```{r eval=T}
neg_dispersion(seq.tab[[6]], positive_sorted_sample = "pos", negative_sorted_sample = "neg1", second_negative_sample = "neg2", type = "superposed")

neg_dispersion(seq.tab[[6]], positive_sorted_sample = "pos", negative_sorted_sample = "neg1", second_negative_sample = "neg2",type = "facet all three")
```

```{r}
# run the following if you want every samples, make sure to change the number of cores
# pdf("test.pdf", width = 10, height = 7)
# mclapply(seq, neg_dispersion, mc.cores = 6, type="superposed")
# dev.off()
```

</div>
## Run the slide_z function
<div style="text-align: justify">
The first step will be to create log2 ratios and log10 abundance (log10 abundance of **pos** \* **neg1** for example) for each ASV between the **pos** and the **neg1** (log_2\_ratio - log10_abundance) and between the ***neg1*** and **neg2** (log2_neg_ratio - log10_neg_abundance). This will be done by the function `log_ratio` called by the `slide_z` function.\
The function will also create an ellipse of confidence interval of your choosing, default being `confidence_interval=c(0.95,0.99, 0.999)`, this is done by the function `ellipse_me` also called by the \`slide_z\` function.

The output will be a list of S4 objects containing the following slots :

-   ig_seq_all : containing all the samples and all the ASV

-   ig_up : containing all the samples and all the ASV that significantly enriched in the **IgA positive fraction**

-   ig_down : containing all the samples and all the ASV that are significantly enriched in the **IgA negative fraction**

In each slot you will find the following columns :

-   taxonomy

-   sample_id

-   new : the sample_data collapsed

-   pos : positive fraction abundance

-   neg1 : negative (9/10) fraction abundance

-   neg2 : negative (1/10) fraction abundance

-   log10_abundance = log10(pos \* neg1)

-   log2_ratio = log2(pos / neg1)

-   log10_neg_abundance = log10(neg1 \* neg2)

-   log2_neg_ratio = log2(neg1 / neg2)

-   taxa = tax_table collapsed

-   SlideNorm = the normalized dispersion for each ASV

-   score = IgAseq score

-   ellipse_level = the level of confidence

```{r eval=T}
IgA_seq=list()

system.time(for(i in names(seq.tab)){
  print(i)
  IgA_seq[[i]]= slide_z(seq.tab[[i]], positive_sorted_sample = "pos", negative_sorted_sample = "neg1", second_negative_sample = "neg2", deltaX = 30, slide_version = "slide_z_modern", alpha = 0.05, plot = F, zero_treatment = "random generation") 
})
```
</div>
## Collapse the list of IgAseq into a single S4 object
<div style="text-align: justify">
We now need to collapse the list of IgA_seq objects into a single one, just use `collapse_IgAseq` and separate the columns that were pasted together.

```{r eval=T, warning=F}
IgA_seq = collapse_IgAseq(IgA_seq)
# DT:: datatable(IgA_seq@ig_seq_all, rownames = F)
# View(IgA_seq)


IgA_all= IgA_seq@ig_seq_all  %>% 
  separate(taxonomy, into=c("Reign","Phylum", "Order","Class","Family", "Genus","Species","ASV", "rest"), sep="#") %>% 
  separate(col = new, into=colnames(sample_data(igaseq)),sep= "#" )

IgA_all$alpha= ifelse(IgA_all$score>1.96, "positively significative", ifelse(IgA_all$score< -1.96, "negatively significative", "not significative"))

dim(IgA_all)
```
</div>
## Plot like Gordon IgA seq
<div style="text-align: justify">
For the example we will plot the IgAseq data like Gordon's paper. We first make a wilcoxon test to decipher which genera are significantly different from zero. Then we plot as balloon plot using ggplot.
</div>
```{r eval=T}
library(rstatix)
tmp= IgA_all %>% 
  group_by(Genus, donor)%>% 
   mutate(n=n())%>%
  filter(n>5 , Genus!="NA")%>%
  wilcox_test(score~1, mu=0, alternative ="two.sided", detailed = T)

tmp2= IgA_all %>%
  group_by(Genus, donor)%>%
  mutate(n=n())%>%
  filter(n>5, Genus!="NA")%>%
  dplyr::select(Genus, score, donor)%>%
    summarise(mean_score= mean(score))%>%
  left_join(tmp)
  alpha= ifelse(tmp2$p<0.05, -log10(tmp2$p), 0.5)
  
  tmp2$donor= factor(tmp2$donor, levels = c("child_delivery","child_2_months","child_24_months", "mother_24_months"))
  tmp2%>%
  ggplot(aes(donor, Genus))+
  geom_point(aes(size=alpha, fill=mean_score), alpha=alpha , shape=21)+
    scale_size_binned(range=c(0,10), trans = "log10")+
    scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0, breaks=c(-15,-10,-5,-2,1,0,1,2,5,10))+
    # scale_x_discrete(labels=c("Delivery", "2 months", "24 months", "Mothers"))+
    labs(fill="IgAseq mean score", x="Age", size="Level of significance")+
       theme(axis.text.y = element_text(size=15, face="bold"),
         axis.text.x=element_text(size=25, face="bold"),
         strip.text.y = element_text(angle = 0, size=15, face="bold"),
         axis.title.x = element_blank(), 
         axis.title.y = element_blank(),
         legend.position = 'bottom',
         legend.title = element_text(size=20, face="bold"), 
         legend.text = element_text(size=20, face="bold"), 
         legend.box = "vertical",
         legend.key.size = unit(2,"cm")
        )

```

## Test models using the IgASeq data

# Use multi-omics models based on `mixOmics` package
