<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Introduction to the ImmuMicrobiome package • ImmuMicrobiome</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css"><script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet"><script src="pkgdown.js"></script><meta property="og:title" content="Introduction to the ImmuMicrobiome package"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-title-body">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ImmuMicrobiome</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="articles/ImmuMicrobiome_vignette.html">ImmuMicrobiome_vignette</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Introduction to the ImmuMicrobiome package</h1>
    </div>


<div id="introduction-to-the-immumicrobiome-package" class="section level1">

<p>Rémy Villette 2023-10-26</p>
</div>
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a></h1>
<div style="text-align: justify">
<p>This package is based on the code of Dr. Villette and Dr. Larsen. The package is written and maintained by Dr. Villette. This package is meant to facilitate microbiome exploration and ensuring nice plotting.<br>
This package covers :</p>
<ul><li><p>The dada2 pipeline with wrapper functions that ease the processing of multiple projects</p></li>
<li><p>Some plotting functions for beta diversity, heatmap and differential abundance analysis using directly a phyloseq object</p></li>
<li><p>A pipeline for IgASeq analysis</p></li>
</ul></div>
</div>
<div class="section level1">
<h1 id="trim-denoise-and-align-your-sequences-using-dada2-pipeline">Trim, denoise and align your sequences using dada2 pipeline<a class="anchor" aria-label="anchor" href="#trim-denoise-and-align-your-sequences-using-dada2-pipeline"></a></h1>
<div style="text-align: justify">
<p>For convenience this will not be a reproducible example, dada2 takes too long to compute and knit. This part of the tutorial will present a run that we performed in house. The rest of the tutorial will be based on reproducible data.</p>
</div>
<div class="section level2">
<h2 id="get-the-files">Get the files<a class="anchor" aria-label="anchor" href="#get-the-files"></a></h2>
<div style="text-align: justify">
<p>We use here a wrapper function that will create a list of three for pair end :</p>
<ul><li><p>forward files</p></li>
<li><p>reverse files</p></li>
<li><p>names of the files</p></li>
</ul><p>And for single end : -</p>
<ul><li><p>a list of files</p></li>
<li><p>names of the files</p></li>
</ul></div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">f_list</span> <span class="op">=</span> <span class="fu"><a href="reference/list_fastq.html">list_fastq</a></span><span class="op">(</span><span class="st">"/home/bigbeast/Documents/tmp/2022-12 CIMMAP run ELISE"</span>, pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"R1"</span>,</span>
<span>    <span class="st">"R2"</span><span class="op">)</span>, separator <span class="op">=</span> <span class="st">"_"</span>, level <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check that all files are distinct</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">f_list</span>, <span class="va">duplicated</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check that all files exist5</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">f_list</span>, <span class="va">file.exists</span><span class="op">)</span></span>
<span><span class="va">random</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">f_list</span>, <span class="st">"["</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">255</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># tmp= summarise_fastq(random, cores =30, plot = F )</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="check-the-quality-profile">Check the quality profile<a class="anchor" aria-label="anchor" href="#check-the-quality-profile"></a></h2>
<div style="text-align: justify">
<p>We will use the function <code>qc_check</code>. This will take time as the <code>plotQualityProfile</code> isn’t parallelized in dada2. This function will create two plots (for pair end) of <code>n</code> aggregated samples and only one plot if you are using single end.</p>
</div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/qc_check.html">qc_check</a></span><span class="op">(</span><span class="va">flist</span>, n <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="check-the-best-trimming-parameters-for-your-set-of-fastq">Check the best trimming parameters for your set of fastq<a class="anchor" aria-label="anchor" href="#check-the-best-trimming-parameters-for-your-set-of-fastq"></a></h2>
<div class="section level3">
<h3 id="test-optimal-cutting-param">Test optimal cutting param<a class="anchor" aria-label="anchor" href="#test-optimal-cutting-param"></a></h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">f_list</span>, <span class="st">"["</span>, <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">255</span>, <span class="fl">30</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tmp_list</span> <span class="op">=</span> <span class="fu"><a href="reference/filt_list.html">filt_list</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></span>
<span><span class="va">filtered</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cb</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">(</span><span class="fl">300</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">50</span>, by <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span>, m <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co"># Very long be careful try to test combination of trimmings</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">cb</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">filtered</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="reference/filter_fastq.html">filter_fastq</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="va">tmp_list</span>, cutting_param <span class="op">=</span> <span class="va">cb</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span>, cores <span class="op">=</span> <span class="fl">35</span>,</span>
<span>        trimleft <span class="op">=</span> <span class="fl">35</span>, maxEE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># extract the data to a df</span></span>
<span><span class="va">per</span> <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">t</span> <span class="op">=</span> <span class="va">filtered</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>        <span class="fu">mutate</span><span class="op">(</span>per <span class="op">=</span> <span class="va">reads.out</span><span class="op">/</span><span class="va">reads.in</span><span class="op">)</span></span>
<span>    <span class="va">per</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">per</span>, <span class="va">t</span><span class="op">$</span><span class="va">per</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="co"># per$names=rownames(tmp[1])</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">per</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">per</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">cb</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>, <span class="va">cb</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="va">per</span><span class="op">$</span><span class="va">names</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">t</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"transmic and IgA rescue mice percent reads passed depending on cutting parameters.png"</span>,</span>
<span>    width <span class="op">=</span> <span class="fl">600</span>, height <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span>
<span><span class="va">per</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"cut"</span>, values_to <span class="op">=</span> <span class="st">"per"</span>, cols <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">per</span>, x <span class="op">=</span> <span class="va">cut</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">names</span>,</span>
<span>    col <span class="op">=</span> <span class="va">names</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"percentage passed"</span>, x <span class="op">=</span> <span class="st">"Trimming parameters"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="test-optimal-maxee">Test optimal maxEE<a class="anchor" aria-label="anchor" href="#test-optimal-maxee"></a></h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cb</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">(</span><span class="fl">7</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span>, by <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span>, m <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">cb</span> <span class="op">=</span> <span class="va">cb</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cb</span><span class="op">)</span><span class="op">:</span><span class="fl">1</span>, <span class="op">]</span></span>
<span><span class="co"># Very long be careful try to test combination of trimmings</span></span>
<span><span class="co"># registerDoParallel(cl = makeCluster(5))</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">cb</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">filtered</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="reference/filter_fastq.html">filter_fastq</a></span><span class="op">(</span><span class="va">tmp</span>, <span class="va">tmp_list</span>, cutting_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">260</span>, <span class="fl">250</span><span class="op">)</span>, cores <span class="op">=</span> <span class="fl">35</span>,</span>
<span>        trimleft <span class="op">=</span> <span class="fl">35</span>, maxEE <span class="op">=</span> <span class="va">cb</span><span class="op">[</span>, <span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># extract the data to a df</span></span>
<span><span class="va">per</span> <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">t</span> <span class="op">=</span> <span class="va">filtered</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>        <span class="fu">mutate</span><span class="op">(</span>per <span class="op">=</span> <span class="va">reads.out</span><span class="op">/</span><span class="va">reads.in</span><span class="op">)</span></span>
<span>    <span class="va">per</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">per</span>, <span class="va">t</span><span class="op">$</span><span class="va">per</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">per</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">cb</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>, <span class="va">cb</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="va">per</span><span class="op">$</span><span class="va">names</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">t</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/png.html" class="external-link">png</a></span><span class="op">(</span><span class="st">"transmic and IgA rescue mice percent reads passed depending on errors parameters.png"</span>,</span>
<span>    width <span class="op">=</span> <span class="fl">600</span>, height <span class="op">=</span> <span class="fl">400</span><span class="op">)</span></span>
<span><span class="va">per</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"cut"</span>, values_to <span class="op">=</span> <span class="st">"per"</span>, cols <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>y <span class="op">=</span> <span class="va">per</span>, x <span class="op">=</span> <span class="va">cut</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>group <span class="op">=</span> <span class="va">names</span>,</span>
<span>    col <span class="op">=</span> <span class="va">names</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>y <span class="op">=</span> <span class="st">"percentage passed"</span>, x <span class="op">=</span> <span class="st">"Trimming parameters"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/grDevices/dev.html" class="external-link">dev.off</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="filter-and-trim-the-fastq">Filter and trim the fastq<a class="anchor" aria-label="anchor" href="#filter-and-trim-the-fastq"></a></h2>
<div style="text-align: justify">
<p>We now have to remove the bad quality reads and trim the length. You will find a function to create the list of filtered files and one to make the filtered files.</p>
</div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filt</span> <span class="op">=</span> <span class="fu"><a href="reference/filt_list.html">filt_list</a></span><span class="op">(</span><span class="va">f_list</span><span class="op">)</span>  <span class="co"># create the list of filtered files</span></span>
<span></span>
<span><span class="va">filtered</span> <span class="op">=</span> <span class="fu">filterAndTrim</span><span class="op">(</span>fwd <span class="op">=</span> <span class="va">fwd</span>, filt <span class="op">=</span> <span class="va">filtFs</span>, rev <span class="op">=</span> <span class="va">rv</span>, filt.rev <span class="op">=</span> <span class="va">filtRs</span>, truncLen <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">260</span>,</span>
<span>    <span class="fl">240</span><span class="op">)</span>, trimLeft <span class="op">=</span> <span class="fl">25</span>, maxEE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">5</span><span class="op">)</span>, multithread <span class="op">=</span> <span class="fl">45</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="analyse-microbiome-system">Analyse microbiome system<a class="anchor" aria-label="anchor" href="#analyse-microbiome-system"></a></h1>
<div class="section level2">
<h2 id="analysis-based-on-phyloseq-objects">Analysis based on phyloseq objects<a class="anchor" aria-label="anchor" href="#analysis-based-on-phyloseq-objects"></a></h2>
<div class="section level3">
<h3 id="alpha-diversity">Alpha diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a></h3>
<div style="text-align: justify">
<p>We will use the <strong>enterotype</strong> data to explore some of the plotting functions. Let’s start with the beta diversity functions <code>beta_diversity</code> and <code>beta_dispersion</code>.</p>
<p>Alpha diversity is an important facet of microbiome analysis. I’ve created wrapper function to plot either alpha diversity as boxplots or as line plots. This function is rudimentary but allows to plot quickly alpha diversity</p>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"enterotype"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/alpha_diversity.html">alpha_diversity</a></span><span class="op">(</span><span class="va">enterotype</span>, measure <span class="op">=</span> <span class="st">"Shannon"</span>, x <span class="op">=</span> <span class="st">"Enterotype"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>,</span>
<span>    plot_type <span class="op">=</span> <span class="st">"boxplot"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/alpha_diversity.html">alpha_diversity</a></span><span class="op">(</span><span class="va">enterotype</span>, measure <span class="op">=</span> <span class="st">"Shannon"</span>, x <span class="op">=</span> <span class="st">"Enterotype"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>,</span>
<span>    plot_type <span class="op">=</span> <span class="st">"line"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Alpha%20diversity%201-1.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/Alpha%20diversity%201-2.png" style="display: block; margin: auto;"></p>
<div style="text-align: justify">
<p>These plots are compatible with other aspects of ggplot and tidyverse in general. You can use facets or stats and even pipes operator to pass functions before using <code><a href="reference/alpha_diversity.html">alpha_diversity()</a></code>. Stats are also implemented in this function but are very sparse, it can be a good idea to make the stat on your own wit <code>stat_compare_means()</code> or <code>stat_pvalue_manual()</code>.</p>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/alpha_diversity.html">alpha_diversity</a></span><span class="op">(</span><span class="va">enterotype</span>, measure <span class="op">=</span> <span class="st">"Shannon"</span>, x <span class="op">=</span> <span class="st">"Enterotype"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>,</span>
<span>    plot_type <span class="op">=</span> <span class="st">"boxplot"</span>, stat <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">enterotype</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">subset_samples</span><span class="op">(</span><span class="va">Enterotype</span> <span class="op">!=</span> <span class="st">"NA"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="reference/alpha_diversity.html">alpha_diversity</a></span><span class="op">(</span>measure <span class="op">=</span> <span class="st">"Shannon"</span>, x <span class="op">=</span> <span class="st">"Enterotype"</span>, group <span class="op">=</span> <span class="st">"Enterotype"</span>,</span>
<span>        plot_type <span class="op">=</span> <span class="st">"boxplot"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">stat_compare_means</span><span class="op">(</span>comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"2"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span><span class="va">SeqTech</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Alpha%20diversity%202-1.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/Alpha%20diversity%202-2.png" style="display: block; margin: auto;"></p>
<div style="text-align: justify">
<p>There is another feature that can become handy at some points, checking the link between depth and alpha diversity. For this, the function allows you to specify <code>check_depth=T</code>. This will produce two plots:</p>
<ul><li>one boxplot</li>
<li>one dotplot with alpha diversity measure as Y and the sum of reads for each sample as X</li>
</ul></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span><span class="op">)</span></span>
<span><span class="va">GlobalPatterns</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="reference/alpha_diversity.html">alpha_diversity</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"SampleType"</span>, group <span class="op">=</span> <span class="st">"SampleType"</span>, check_depth <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Alpha%20diversity%203-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="beta-diversity">Beta diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a></h3>
<div style="text-align: justify">
<p>You will have the choice between <code>beta_diversity</code> and <code>beta_dispersion</code> for your beta diversity plotting.</p>
<p><code>beta_dispersion</code> will plot PCoA, NMDS, PCA, DCA, CA and t-SNE for the moment. This function will plot the two components of your choosing, <strong>confidence ellipses</strong> and boxplot for each axis and for each group. Each function will return a plot and a percentage of contribution for each component.</p>
<p><code>beta_diversity</code> will be removed progressively from this package. This function is redundant with <code>beta_dispersion</code>.</p>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/beta_diversity.html">beta_diversity</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"PCoA"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>, permanova <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"PCoA"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%201-1.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%201-2.png" style="display: block; margin: auto;"></p>
<p>The <code><a href="reference/beta_dispersion.html">beta_dispersion()</a></code> function comes with three choices of plotting:</p>
<ul><li>Boxplot on the side</li>
<li>Just the multidimensional reduction</li>
<li>Multidimensional reduction and loadings if the algorithm allows it</li>
</ul><p>NMDS, PCA, CA, DCA are compatible with the <code>type="arrows"</code> argument while PCoA is not. Arrows can be tricky to use with all the taxa in a phyloseq object. This plotting option is more interesting with sparse data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Boxplots on the sides</span></span>
<span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"PCoA"</span>, type <span class="op">=</span> <span class="st">"boxplot"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>,</span>
<span>    stat <span class="op">=</span> <span class="st">"permanova"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>, <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>, text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Nothing</span></span>
<span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"PCoA"</span>, type <span class="op">=</span> <span class="st">"pure"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>,</span>
<span>    stat <span class="op">=</span> <span class="st">"permanova"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>, <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>, text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Arrows</span></span>
<span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"PCA"</span>, type <span class="op">=</span> <span class="st">"arrows"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>,</span>
<span>    stat <span class="op">=</span> <span class="st">"permanova"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>, <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">0.1</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>, text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%202-1.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%202-2.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%202-3.png" style="display: block; margin: auto;"></p>
<div class="section level4">
<h4 id="multiple-dimension-reductions">Multiple dimension reductions<a class="anchor" aria-label="anchor" href="#multiple-dimension-reductions"></a></h4>
<div style="text-align: justify">
<p>There are unconstrained and constrained approaches to beta diversity. Unconstrained analysis, more common in literature, are either based on:</p>
<ul><li>dissimilarity matrix (PCoA, NMDS, DCA)</li>
<li>directly on the raw matrix (PCA, CA)</li>
<li>t-SNE that can use both dissimilarity matrix or the raw matrix</li>
</ul></div>
<div class="section level5">
<h5 id="unconstrained-models">Unconstrained models<a class="anchor" aria-label="anchor" href="#unconstrained-models"></a></h5>
<p>You can play with the parameters of each function like the following plots :</p>
<div class="section level6">
<h6 id="pcoa">PCoA<a class="anchor" aria-label="anchor" href="#pcoa"></a></h6>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"PCoA"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>, stat <span class="op">=</span> <span class="st">"permanova"</span>,</span>
<span>    color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>, <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>, text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%203-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="t-sne">t-SNE<a class="anchor" aria-label="anchor" href="#t-sne"></a></h6>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"tsne"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>,</span>
<span>    <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, stat <span class="op">=</span> <span class="st">"permanova"</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech"</span>,</span>
<span>    lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>, text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span>, where <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%204-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="nmds">NMDS<a class="anchor" aria-label="anchor" href="#nmds"></a></h6>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"NMDS"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>,</span>
<span>    <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech"</span>, lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>,</span>
<span>    text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span>, where <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Run 0 stress 0.1380077 </span></span>
<span><span class="co">#&gt; Run 1 stress 0.1536597 </span></span>
<span><span class="co">#&gt; Run 2 stress 0.1596971 </span></span>
<span><span class="co">#&gt; Run 3 stress 0.1592765 </span></span>
<span><span class="co">#&gt; Run 4 stress 0.1499595 </span></span>
<span><span class="co">#&gt; Run 5 stress 0.1582167 </span></span>
<span><span class="co">#&gt; Run 6 stress 0.1549281 </span></span>
<span><span class="co">#&gt; Run 7 stress 0.15729 </span></span>
<span><span class="co">#&gt; Run 8 stress 0.1578031 </span></span>
<span><span class="co">#&gt; Run 9 stress 0.1521635 </span></span>
<span><span class="co">#&gt; Run 10 stress 0.148488 </span></span>
<span><span class="co">#&gt; Run 11 stress 0.150579 </span></span>
<span><span class="co">#&gt; Run 12 stress 0.1458337 </span></span>
<span><span class="co">#&gt; Run 13 stress 0.1539079 </span></span>
<span><span class="co">#&gt; Run 14 stress 0.1443015 </span></span>
<span><span class="co">#&gt; Run 15 stress 0.1556298 </span></span>
<span><span class="co">#&gt; Run 16 stress 0.1508059 </span></span>
<span><span class="co">#&gt; Run 17 stress 0.1549227 </span></span>
<span><span class="co">#&gt; Run 18 stress 0.1603283 </span></span>
<span><span class="co">#&gt; Run 19 stress 0.1614954 </span></span>
<span><span class="co">#&gt; Run 20 stress 0.153975 </span></span>
<span><span class="co">#&gt; *** Best solution was not repeated -- monoMDS stopping criteria:</span></span>
<span><span class="co">#&gt;     18: stress ratio &gt; sratmax</span></span>
<span><span class="co">#&gt;      2: scale factor of the gradient &lt; sfgrmin</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%205-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="dca">DCA<a class="anchor" aria-label="anchor" href="#dca"></a></h6>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"DCA"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>,</span>
<span>    <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech"</span>, lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>,</span>
<span>    text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span>, where <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%206-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a></h6>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"PCA"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>,</span>
<span>    <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech without boxplots"</span>, lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>, text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span>, where <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%207-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="ca">CA<a class="anchor" aria-label="anchor" href="#ca"></a></h6>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, dist <span class="op">=</span> <span class="st">"bray"</span>, method <span class="op">=</span> <span class="st">"CA"</span>, group <span class="op">=</span> <span class="st">"SeqTech"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>,</span>
<span>    <span class="st">"#117777"</span>, <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Sequencing tech without boxplots"</span>, lwd <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"polygon"</span>, text <span class="op">=</span> <span class="cn">T</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span>, where <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%208-1.png" style="display: block; margin: auto;"> ##### {-}</p>
</div>
</div>
</div>
<div class="section level4">
<h4 id="constrained-models">Constrained models<a class="anchor" aria-label="anchor" href="#constrained-models"></a></h4>
<p>Constrained analysis are multiple regression of the unconstrained approaches. They allow to fit the ecological data or sample data to the community data. Alternatively, you can fit other omic data to your beta diversity analysis. We have two different classes of constrained analysis :</p>
<ul><li>
<code>rda</code> is a regression of a PCA</li>
<li>
<code>cca</code> is a regression of a CA</li>
<li>
<code>dbRDA</code> is a RDA based on a dissimilarity matrix<br>
The function will return a plot, same as <code>beta_dispersion</code>, and also the result of the constrained analysis.</li>
</ul><p>Vegan authors have implemented a very nice feature if you want to make your own representation instead of using mine. You can set use <code>scores(res, tidy=T)</code> to return a dataframe compatible with ggplot2.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mod</span> <span class="op">=</span> <span class="st">"SeqTech+Gender+Nationality+Age+ClinicalStatus"</span></span></code></pre></div>
<div class="section level6">
<h6 id="cca">CCA<a class="anchor" aria-label="anchor" href="#cca"></a></h6>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="reference/constrained_beta_dispersion.html">constrained_beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, model <span class="op">=</span> <span class="va">mod</span>, group <span class="op">=</span> <span class="st">"Enterotype"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"CCA"</span>, text <span class="op">=</span> <span class="cn">T</span>, color_vector <span class="op">=</span> <span class="va">tol21rainbow</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%209-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="rda">RDA<a class="anchor" aria-label="anchor" href="#rda"></a></h6>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="reference/constrained_beta_dispersion.html">constrained_beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, model <span class="op">=</span> <span class="va">mod</span>, group <span class="op">=</span> <span class="st">"Enterotype"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"RDA"</span>, text <span class="op">=</span> <span class="cn">T</span>, color_vector <span class="op">=</span> <span class="va">tol21rainbow</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%2010-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level6">
<h6 id="dbrda">dbRDA<a class="anchor" aria-label="anchor" href="#dbrda"></a></h6>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="reference/constrained_beta_dispersion.html">constrained_beta_dispersion</a></span><span class="op">(</span><span class="va">enterotype</span>, model <span class="op">=</span> <span class="va">mod</span>, group <span class="op">=</span> <span class="st">"Enterotype"</span>,</span>
<span>    method <span class="op">=</span> <span class="st">"dbRDA"</span>, text <span class="op">=</span> <span class="cn">T</span>, color_vector <span class="op">=</span> <span class="va">tol21rainbow</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Beta%20diversity%2011-1.png" style="display: block; margin: auto;"></p>
<p>You can then use <code>anova</code> statistic test to decipher which factor or feature is significantly associated with your beta diversity.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html" class="external-link">anova</a></span><span class="op">(</span><span class="va">res</span>, by <span class="op">=</span> <span class="st">"terms"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Permutation test for capscale under reduced model</span></span>
<span><span class="co">#&gt; Terms added sequentially (first to last)</span></span>
<span><span class="co">#&gt; Permutation: free</span></span>
<span><span class="co">#&gt; Number of permutations: 999</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Model: capscale(formula = as(otu_table(reverseASV(physeq)), "matrix") ~ SeqTech + Gender + Nationality + Age + ClinicalStatus, data = df, distance = dist, na.action = na.exclude)</span></span>
<span><span class="co">#&gt;                Df SumOfSqs      F Pr(&gt;F)    </span></span>
<span><span class="co">#&gt; Gender          1  0.06650 0.7390  0.639    </span></span>
<span><span class="co">#&gt; Nationality     5  1.06616 2.3695  0.005 ** </span></span>
<span><span class="co">#&gt; Age             1  0.43786 4.8657  0.001 ***</span></span>
<span><span class="co">#&gt; ClinicalStatus  3  0.30763 1.1395  0.291    </span></span>
<span><span class="co">#&gt; Residual       26  2.33975                  </span></span>
<span><span class="co">#&gt; ---</span></span>
<span><span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="heatmap-based-on-the-phyloseq-object">Heatmap based on the phyloseq object<a class="anchor" aria-label="anchor" href="#heatmap-based-on-the-phyloseq-object"></a></h3>
<div style="text-align: justify">
<p>This function will perform a top taxa at a rank of your choosing and create a heatmap with annotations. For now only one annotation is supported.<br>
The clustering is made using the <code>hclust</code>function and a Ward.D2 method. You can define the distance matrix that you want to use. It is important to note that the distance is made before any trimming of the data. This means that the distance matrix is made at the ASV/OTU level before doing the rank merging and topping, so the clustering will represent your “true” data instead of a modified dataset. In my personnal opinion, this is the only way of performing clustering or any type of generalized analysis : clustering, reducing or else on the ASV/OTU/MAG/… level and then aggregates for plotting.</p>
<p>If you really want to cluster on other taxonomic level you should use <code>tax_glom</code> before using this function.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/phylo_heatmap.html">phylo_heatmap</a></span><span class="op">(</span><span class="va">enterotype</span>, top <span class="op">=</span> <span class="fl">30</span>, labels <span class="op">=</span> <span class="st">"SeqTech"</span>, taxa_rank <span class="op">=</span> <span class="st">"Genus"</span>, factor_to_plot <span class="op">=</span> <span class="st">"Enterotype"</span>,</span>
<span>    split <span class="op">=</span> <span class="fl">3</span>, distance <span class="op">=</span> <span class="st">"bray"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "No phylogenetic tree in this phyloseq object, bray-curtis distance selected."</span></span>
<span><span class="co">#&gt; [1] " not reversed"</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Phylo%20heatmap%201-1.png" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="usage-of-various-functions">Usage of various functions<a class="anchor" aria-label="anchor" href="#usage-of-various-functions"></a></h2>
<div style="text-align: justify">
<p>The idea behind these functions is : creating a more automatic pipeline enabling filtering and subseting of the <code>phyloseq</code> object without having to perform a temporary <code>phyloseq</code> object and a temporary distance object. With these function you can directly use the “pipe” introduced by <code>magrittr</code>.<br>
So you can use a <code>subset_samples</code> like the following and automatically plot your beta diversity without adding too much code.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enterotype</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">subset_samples</span><span class="op">(</span><span class="va">SeqTech</span> <span class="op">==</span> <span class="st">"Illumina"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span>group <span class="op">=</span> <span class="st">"Enterotype"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>, <span class="st">"#117777"</span>,</span>
<span>        <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Enterotypes \n with bray"</span>, lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">2</span>,</span>
<span>        draw <span class="op">=</span> <span class="st">"lines"</span>, text <span class="op">=</span> <span class="cn">T</span>, stat <span class="op">=</span> <span class="st">"permanova"</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span>, where <span class="op">=</span> <span class="st">"bottomleft"</span>,</span>
<span>        cex <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/Phylo%20heatmap%202-1.png" style="display: block; margin: auto;"></p>
<p>Additionally, if you want to go further you can also serialized the code with a <code>for</code> loop or a <code>lapply</code> or even a parallel approach using <code>mclapply</code> or <code>doParallel</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/layout.html" class="external-link">layout</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span>, ncol <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Illumina"</span>, <span class="st">"Sanger"</span>, <span class="st">"Pyro454"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">enterotype</span> <span class="op">%&gt;%</span></span>
<span>        <span class="fu">subset_samples</span><span class="op">(</span><span class="va">SeqTech</span> <span class="op">==</span> <span class="va">i</span> <span class="op">&amp;</span> <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">Enterotype</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>        <span class="fu"><a href="reference/beta_dispersion.html">beta_dispersion</a></span><span class="op">(</span>group <span class="op">=</span> <span class="st">"Enterotype"</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#777711"</span>, <span class="st">"#117777"</span>,</span>
<span>            <span class="st">"#DD7788"</span><span class="op">)</span>, legend_title <span class="op">=</span> <span class="st">"Enterotypes \n with bray"</span>, type <span class="op">=</span> <span class="st">"pure"</span>,</span>
<span>            lwd <span class="op">=</span> <span class="fl">2</span>, font <span class="op">=</span> <span class="fl">2</span>, draw <span class="op">=</span> <span class="st">"lines"</span>, text <span class="op">=</span> <span class="cn">T</span>, stat <span class="op">=</span> <span class="st">"none"</span>, y.intersp <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>            where <span class="op">=</span> <span class="st">"bottomleft"</span>, cex <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-8-1.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-8-2.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-8-3.png" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="differential-abundance-testing">Differential abundance testing<a class="anchor" aria-label="anchor" href="#differential-abundance-testing"></a></h2>
<p>Differential abundance testing is quite complicated because of the number of different statistical approaches existing in the literature. I personally use ALDEx2 and SIAMCAT a lot. The first because it is usually highly regarded and the second for the overall easiness of the package. For now <code>differential_abundance</code> implement only <code>ALDEx2</code> approach.</p>
<p>The function will return two datasets (all taxa and one with only the significant taxa) and two plots representing the taxa significantly represented in one of the two conditions. This function only covers two by two analysis.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">=</span> <span class="va">GlobalPatterns</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">subset_samples</span><span class="op">(</span><span class="va">SampleType</span> <span class="op">==</span> <span class="st">"Feces"</span> <span class="op">|</span> <span class="va">SampleType</span> <span class="op">==</span> <span class="st">"Soil"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">subset_taxa</span><span class="op">(</span><span class="va">Genus</span> <span class="op">!=</span> <span class="st">"NA"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">tax_glom</span><span class="op">(</span><span class="st">"Genus"</span><span class="op">)</span></span>
<span><span class="fu">taxa_names</span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">tax_table</span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span><span class="op">[</span>, <span class="st">"Genus"</span><span class="op">]</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu">tax_table</span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span><span class="op">[</span>, <span class="st">"Genus"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="va">tmp</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="reference/differential_abundance.html">differential_abundance</a></span><span class="op">(</span>group <span class="op">=</span> <span class="st">"SampleType"</span>, col1 <span class="op">=</span> <span class="st">"brown"</span>, col2 <span class="op">=</span> <span class="st">"darkgreen"</span>,</span>
<span>        plot <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="results">Results<a class="anchor" aria-label="anchor" href="#results"></a></h3>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">all_features</span><span class="op">)</span></span>
<span><span class="co">#&gt;                                 we.ep     we.eBH      wi.ep    wi.eBH</span></span>
<span><span class="co">#&gt; Nitrosopumilus3           0.038020857 0.12553457 0.07321429 0.1739762</span></span>
<span><span class="co">#&gt; CandidatusNitrososphaera4 0.080883984 0.21608822 0.05714286 0.1534536</span></span>
<span><span class="co">#&gt; Methanocorpusculum15      0.393517314 0.52330393 0.50937500 0.6131738</span></span>
<span><span class="co">#&gt; Methanobacterium17        0.192887050 0.32809281 0.22232143 0.3464315</span></span>
<span><span class="co">#&gt; Methanobrevibacter19      0.005763966 0.04577142 0.05714286 0.1534536</span></span>
<span><span class="co">#&gt; Propionibacterium20       0.575347882 0.69177092 0.69397321 0.7805085</span></span>
<span><span class="co">#&gt;                               rab.all rab.win.Feces rab.win.Soil    diff.btw</span></span>
<span><span class="co">#&gt; Nitrosopumilus3           -0.03578414     2.1640049  -3.54657967  -6.0725458</span></span>
<span><span class="co">#&gt; CandidatusNitrososphaera4  2.94274075     1.7275926   6.88742882   5.5007228</span></span>
<span><span class="co">#&gt; Methanocorpusculum15      -0.74114455    -1.6164717  -0.03037559   1.3795403</span></span>
<span><span class="co">#&gt; Methanobacterium17        -1.58879597     0.6563627  -3.78683212  -4.4195807</span></span>
<span><span class="co">#&gt; Methanobrevibacter19       5.47119787     9.1212605  -3.52760211 -13.2996558</span></span>
<span><span class="co">#&gt; Propionibacterium20       -2.23757862    -1.6605794  -3.39678481  -0.9666516</span></span>
<span><span class="co">#&gt;                           diff.win     effect      overlap</span></span>
<span><span class="co">#&gt; Nitrosopumilus3           3.082815 -1.8944361 0.0155616731</span></span>
<span><span class="co">#&gt; CandidatusNitrososphaera4 3.935999  1.8209332 0.0003653998</span></span>
<span><span class="co">#&gt; Methanocorpusculum15      3.092592  0.4041233 0.2953373376</span></span>
<span><span class="co">#&gt; Methanobacterium17        4.154777 -0.8932739 0.1398980468</span></span>
<span><span class="co">#&gt; Methanobrevibacter19      4.791323 -2.9130433 0.0003653998</span></span>
<span><span class="co">#&gt; Propionibacterium20       5.166144 -0.1619715 0.4322918226</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="barplot-results">Barplot results<a class="anchor" aria-label="anchor" href="#barplot-results"></a></h3>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">$</span><span class="va">barplot</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-11-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="volcano-plot-results">Volcano plot results<a class="anchor" aria-label="anchor" href="#volcano-plot-results"></a></h3>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span><span class="op">$</span><span class="va">volcano</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-12-1.png" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="non-phyloseq-object-oriented-functions">Non phyloseq object oriented functions<a class="anchor" aria-label="anchor" href="#non-phyloseq-object-oriented-functions"></a></h2>
<p>::: {style=“text-align: justify”} To enforce same plotting capacity on non phyloseq objects, the <code>plot_reduction</code> and <code>plot_constrained_reduction</code> have been created. These functions needs a data.frame containing categorical data and the numeric values to make the reduction. These functions are perfect to analyse metabolomic data, flow cytometry data or else. Usage remains the same, with the same arguments than in <code>beta_dispersion</code> and <code>constrained_beta_dispersion</code>.</p>
<div class="section level3">
<h3 id="reduction-of-dimensions-on-non-phyloseq-objects">Reduction of dimensions on non phyloseq objects<a class="anchor" aria-label="anchor" href="#reduction-of-dimensions-on-non-phyloseq-objects"></a></h3>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_reduction.html">plot_reduction</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_constrained_reduction.html">plot_constrained_reduction</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>The first function is an equivalent of and the second one is an equivalent of . Make sure to use data.frame with samples as rows.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">metabolomic</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_reduction.html">plot_reduction</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">metabolomic</span>, clinical_data <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, group <span class="op">=</span> <span class="st">"birth_type"</span>, method <span class="op">=</span> <span class="st">"CA"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"pure"</span>, stat <span class="op">=</span> <span class="st">"permanova"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_reduction.html">plot_reduction</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">metabolomic</span>, clinical_data <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, group <span class="op">=</span> <span class="st">"birth_type"</span>, method <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"pure"</span>, stat <span class="op">=</span> <span class="st">"permanova"</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_reduction.html">plot_reduction</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">metabolomic</span>, clinical_data <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, group <span class="op">=</span> <span class="st">"birth_type"</span>, method <span class="op">=</span> <span class="st">"PCoA"</span>,</span>
<span>    dist <span class="op">=</span> <span class="st">"euclidean"</span>, type <span class="op">=</span> <span class="st">"pure"</span>, stat <span class="op">=</span> <span class="st">"permanova"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/reduction2-1.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/reduction2-2.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/reduction2-3.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="pls-da">PLS-DA<a class="anchor" aria-label="anchor" href="#pls-da"></a></h3>
<div style="text-align: justify">
<p>Partial Least Square Discriminant Analysis are widely used, especially in metabolomics. This particular statistical approach is not yet supported in <code>constrained_beta_dispersion</code>and <code>plot_constrained_reduction</code> but will be in the late versions. For now, this function will use the results from the function and use ggplot2 to make the graph. You can customize the graph using classic <code>ggplot2</code> functions.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">results</span> <span class="op">=</span> <span class="fu">mixOmics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/mixOmics/man/plsda.html" class="external-link">plsda</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">metabolomic</span><span class="op">[</span>, <span class="fl">5</span><span class="op">:</span><span class="fl">217</span><span class="op">]</span>, Y <span class="op">=</span> <span class="va">metabolomic</span><span class="op">$</span><span class="va">birth_type</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># basic graphic</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_plsda.html">plot_plsda</a></span><span class="op">(</span><span class="va">results</span>, top.loads <span class="op">=</span> <span class="fl">20</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown"</span>, <span class="st">"orange"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># with a bit of improvement</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_plsda.html">plot_plsda</a></span><span class="op">(</span><span class="va">results</span>, top.loads <span class="op">=</span> <span class="fl">20</span>, color_vector <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown"</span>, <span class="st">"orange"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span><span class="st">"Birth route"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>, axis.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">15</span>,</span>
<span>    face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, panel.grid <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>,</span>
<span>    legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">20</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, axis.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/plsda1-1.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/plsda1-2.png" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level3">
<h3 id="analysis-of-variance">Analysis of variance<a class="anchor" aria-label="anchor" href="#analysis-of-variance"></a></h3>
<div style="text-align: justify">
<p>Here is an important part of any analysis : finding which factors explains the most variance. The principle is that we will get the variance for each features and the variance for each group. The group variance will be summed for the given factor and then we will use 1-variance.factor/variance.total, which will give us the variance for each factors.</p>
<p>On top of the variance, the function will calculate p.values using kruskal tests or wilcoxon test depending of the number of factors and a FDR correction. For now the function only supports non parametric tests.</p>
<p>The function can take quite some time if you have a lot of samples, features and factors, you can use multi-threading here.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">var</span> <span class="op">=</span> <span class="va">metabolomic</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">!</span><span class="va">child_id</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="reference/calculate_variance.html">calculate_variance</a></span><span class="op">(</span>clinical_data <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">var</span>, <span class="va">head</span>, <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; $variance</span></span>
<span><span class="co">#&gt; # A tibble: 5 × 5</span></span>
<span><span class="co">#&gt;   features               variable   birth_type breastfeeding          sex</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                  &lt;fct&gt;           &lt;dbl&gt;         &lt;dbl&gt;        &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 1,5-Naphthalenediamine var.tot  0.0000000137  0.0000000137 0.0000000137</span></span>
<span><span class="co">#&gt; 2 1,7-Dimethyluric acid  var.tot  0.0000000855  0.0000000855 0.0000000855</span></span>
<span><span class="co">#&gt; 3 1-Methyladenine        var.tot  0.0000000440  0.0000000440 0.0000000440</span></span>
<span><span class="co">#&gt; 4 1-Methylguanine        var.tot  0.000000345   0.000000345  0.000000345 </span></span>
<span><span class="co">#&gt; 5 1-Vinylimidazole       var.tot  0.000000394   0.000000394  0.000000394 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $p.value</span></span>
<span><span class="co">#&gt; # A tibble: 5 × 4</span></span>
<span><span class="co">#&gt;   features                    p p.adj factor    </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                   &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;     </span></span>
<span><span class="co">#&gt; 1 1,5-Naphthalenediamine 0.0993 0.414 birth_type</span></span>
<span><span class="co">#&gt; 2 1,7-Dimethyluric acid  0.609  0.786 birth_type</span></span>
<span><span class="co">#&gt; 3 1-Methyladenine        0.0555 0.404 birth_type</span></span>
<span><span class="co">#&gt; 4 1-Methylguanine        0.266  0.596 birth_type</span></span>
<span><span class="co">#&gt; 5 1-Vinylimidazole       0.64   0.807 birth_type</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $mean.feat</span></span>
<span><span class="co">#&gt;                 features    mean.feat</span></span>
<span><span class="co">#&gt; 1 1,5-Naphthalenediamine 0.0001591865</span></span>
<span><span class="co">#&gt; 2  1,7-Dimethyluric acid 0.0002805524</span></span>
<span><span class="co">#&gt; 3        1-Methyladenine 0.0001991781</span></span>
<span><span class="co">#&gt; 4        1-Methylguanine 0.0003367005</span></span>
<span><span class="co">#&gt; 5       1-Vinylimidazole 0.0001454276</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $all</span></span>
<span><span class="co">#&gt; $all$birth_type</span></span>
<span><span class="co">#&gt; # A tibble: 213 × 5</span></span>
<span><span class="co">#&gt;    features                        var.tot mean.feat  var.grp var.exp</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                             &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 1,5-Naphthalenediamine         1.37e- 8 0.000159  1.27e- 8  0.0768</span></span>
<span><span class="co">#&gt;  2 1,7-Dimethyluric acid          8.55e- 8 0.000281  8.44e- 8  0.0130</span></span>
<span><span class="co">#&gt;  3 1-Methyladenine                4.40e- 8 0.000199  4.27e- 8  0.0313</span></span>
<span><span class="co">#&gt;  4 1-Methylguanine                3.45e- 7 0.000337  3.37e- 7  0.0238</span></span>
<span><span class="co">#&gt;  5 1-Vinylimidazole               3.94e- 7 0.000145  3.88e- 7  0.0158</span></span>
<span><span class="co">#&gt;  6 17α-Hydroxyprogesterone        6.76e- 8 0.000156  6.64e- 8  0.0169</span></span>
<span><span class="co">#&gt;  7 1H-indene-3-carboxamide        8.47e-10 0.0000441 8.12e-10  0.0411</span></span>
<span><span class="co">#&gt;  8 2,3-pyridinecarboxylic acid    1.25e- 7 0.000426  1.23e- 7  0.0125</span></span>
<span><span class="co">#&gt;  9 2-(hydroxymethyl)butanoic acid 2.06e- 6 0.000646  2.02e- 6  0.0150</span></span>
<span><span class="co">#&gt; 10 2-Fucosyllactose               1.50e- 5 0.00142   1.43e- 5  0.0482</span></span>
<span><span class="co">#&gt; # ℹ 203 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $all$breastfeeding</span></span>
<span><span class="co">#&gt; # A tibble: 213 × 5</span></span>
<span><span class="co">#&gt;    features                        var.tot mean.feat  var.grp var.exp</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                             &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 1,5-Naphthalenediamine         1.37e- 8 0.000159  1.31e- 8  0.0464</span></span>
<span><span class="co">#&gt;  2 1,7-Dimethyluric acid          8.55e- 8 0.000281  8.44e- 8  0.0134</span></span>
<span><span class="co">#&gt;  3 1-Methyladenine                4.40e- 8 0.000199  4.35e- 8  0.0126</span></span>
<span><span class="co">#&gt;  4 1-Methylguanine                3.45e- 7 0.000337  3.40e- 7  0.0154</span></span>
<span><span class="co">#&gt;  5 1-Vinylimidazole               3.94e- 7 0.000145  3.89e- 7  0.0117</span></span>
<span><span class="co">#&gt;  6 17α-Hydroxyprogesterone        6.76e- 8 0.000156  6.56e- 8  0.0291</span></span>
<span><span class="co">#&gt;  7 1H-indene-3-carboxamide        8.47e-10 0.0000441 8.13e-10  0.0404</span></span>
<span><span class="co">#&gt;  8 2,3-pyridinecarboxylic acid    1.25e- 7 0.000426  1.23e- 7  0.0108</span></span>
<span><span class="co">#&gt;  9 2-(hydroxymethyl)butanoic acid 2.06e- 6 0.000646  2.03e- 6  0.0113</span></span>
<span><span class="co">#&gt; 10 2-Fucosyllactose               1.50e- 5 0.00142   1.48e- 5  0.0139</span></span>
<span><span class="co">#&gt; # ℹ 203 more rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $all$sex</span></span>
<span><span class="co">#&gt; # A tibble: 213 × 5</span></span>
<span><span class="co">#&gt;    features                        var.tot mean.feat  var.grp var.exp</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                             &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 1,5-Naphthalenediamine         1.37e- 8 0.000159  1.33e- 8  0.0303</span></span>
<span><span class="co">#&gt;  2 1,7-Dimethyluric acid          8.55e- 8 0.000281  8.46e- 8  0.0106</span></span>
<span><span class="co">#&gt;  3 1-Methyladenine                4.40e- 8 0.000199  3.92e- 8  0.109 </span></span>
<span><span class="co">#&gt;  4 1-Methylguanine                3.45e- 7 0.000337  3.39e- 7  0.0192</span></span>
<span><span class="co">#&gt;  5 1-Vinylimidazole               3.94e- 7 0.000145  3.83e- 7  0.0270</span></span>
<span><span class="co">#&gt;  6 17α-Hydroxyprogesterone        6.76e- 8 0.000156  6.64e- 8  0.0174</span></span>
<span><span class="co">#&gt;  7 1H-indene-3-carboxamide        8.47e-10 0.0000441 8.27e-10  0.0230</span></span>
<span><span class="co">#&gt;  8 2,3-pyridinecarboxylic acid    1.25e- 7 0.000426  1.23e- 7  0.0102</span></span>
<span><span class="co">#&gt;  9 2-(hydroxymethyl)butanoic acid 2.06e- 6 0.000646  2.01e- 6  0.0224</span></span>
<span><span class="co">#&gt; 10 2-Fucosyllactose               1.50e- 5 0.00142   1.48e- 5  0.0172</span></span>
<span><span class="co">#&gt; # ℹ 203 more rows</span></span></code></pre></div>
<p>The function will produce a list of data.frame, one for each factor in your dataset. In each data.frame you will find:</p>
<ul><li>variance: a data.frame with total variance, the group variance, the explained variance and the mean quantity of a given features
<ul><li>“var.tot” aka total variance for the factor, all NA value for the given factor are removed, so the total variance can differ between factors</li>
<li>“var.grp” the variance explained by the factors</li>
<li>“var.exp” explained variance as described by <span class="math display"><em>e</em><em>x</em><em>p</em><em>l</em><em>a</em><em>i</em><em>n</em><em>e</em><em>d</em><em>v</em><em>a</em><em>r</em><em>i</em><em>a</em><em>n</em><em>c</em><em>e</em> = 1 − <em>g</em><em>r</em><em>o</em><em>u</em><em>p</em><em>v</em><em>a</em><em>r</em><em>i</em><em>a</em><em>n</em><em>c</em><em>e</em>/<em>t</em><em>o</em><em>t</em><em>a</em><em>l</em><em>v</em><em>a</em><em>r</em><em>i</em><em>a</em><em>n</em><em>c</em><em>e</em></span>
</li>
<li>“mean.feat” the mean value for a given feature</li>
</ul></li>
<li>p.value : a data.frame with the p values and p values corrected with FDR for a given feature and a given factor. Currently, the function support only non parametric tests.</li>
<li>mean.feat: the mean value for a given feature</li>
</ul></div>
<div class="section level4">
<h4 id="plot-variance">Plot variance<a class="anchor" aria-label="anchor" href="#plot-variance"></a></h4>
<div style="text-align: justify">
<p>Now that the variance is calculated for all your factors, you need to visualise it. I’ve come up with different ways to visualise the variance of a given dataset:</p>
<ul><li><code>plot_all_variance</code></li>
<li><code>plot_xy_variance</code></li>
<li><code>circular_variance_plot</code></li>
</ul></div>
<div class="section level5">
<h5 id="plot-all-variance">Plot all variance<a class="anchor" aria-label="anchor" href="#plot-all-variance"></a></h5>
<div style="text-align: justify">
<p><code>plot_all_variance</code> will give a boxplot graph with a dot for each features, the size is determined by the mean.feat value for a given feature and the color is based on the p.values.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_all_variance.html">plot_all_variance</a></span><span class="op">(</span><span class="va">var</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown"</span>, <span class="st">"darkgreen"</span>, <span class="st">"grey"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/variance%202-1.png" style="display: block; margin: auto;"> Alternatively, the function accepts the <code>heatmap</code>argument, this will produce a basic heatmap. However, this part of the function is still under development at the moment. In the end the function will provide the possibility to clusterise features.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_all_variance.html">plot_all_variance</a></span><span class="op">(</span><span class="va">var</span>, plot_type <span class="op">=</span> <span class="st">"heatmap"</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/variance%203-1.png" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level5">
<h5 id="plot-variance-two-by-two">Plot variance two by two<a class="anchor" aria-label="anchor" href="#plot-variance-two-by-two"></a></h5>
<div style="text-align: justify">
<p><code>plot_xy_variance</code> will return a dot plot with factor 1 as X and factor 2 as Y. The p-values are displayed as following: “Both factors”, “factor 1”, “factor 2”, “None”.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_xy_variance.html">plot_xy_variance</a></span><span class="op">(</span><span class="va">var</span>, x <span class="op">=</span> <span class="st">"birth_type"</span>, y <span class="op">=</span> <span class="st">"breastfeeding"</span>, corrected <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/variance%204-1.png" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level5">
<h5 id="plot-variance-as-circular-plot">Plot variance as circular plot<a class="anchor" aria-label="anchor" href="#plot-variance-as-circular-plot"></a></h5>
<p>Another way to visualise the variance in the dataset is to use the <code>circular_variance_plot</code> function. This function is not really suitable when many features are present.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="reference/circular_variance_plot.html">circular_variance_plot</a></span><span class="op">(</span><span class="va">var</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metabolomic</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">select</span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="reference/calculate_variance.html">calculate_variance</a></span><span class="op">(</span>clinical_data <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu"><a href="reference/circular_variance_plot.html">circular_variance_plot</a></span><span class="op">(</span>adjust <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylim</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.05</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/variance%205-1.png" style="display: block; margin: auto;"><img src="ImmuMicrobiome_book_files/figure-gfm/variance%205-2.png" style="display: block; margin: auto;"></p>
</div>
</div>
</div>
<div class="section level3">
<h3 id="correlations">Correlations<a class="anchor" aria-label="anchor" href="#correlations"></a></h3>
<div style="text-align: justify">
<p>One last aspect of classical biostatisics is correlations. It can be extremely handy to verify correlation between taxa and metabolomics for example. Correlation can be presented as dotplot with a regression line or as heatmaps/correlograms. Good packages for plotting correlogramms already exists on R, I’ve decided to implement a heatmap correlation graph.<br>
For thsi example we will take only 30 random features to ensure plotting capacity.</p>
</div>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># take randomly 30 features from the metabolomic dataset</span></span>
<span><span class="va">rng</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">217</span>, <span class="fl">30</span><span class="op">)</span></span>
<span><span class="va">rng2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fl">5</span><span class="op">:</span><span class="fl">217</span>, <span class="fl">30</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_corr_heatmap.html">plot_corr_heatmap</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">metabolomic</span><span class="op">[</span>, <span class="va">rng</span><span class="op">]</span>, Y <span class="op">=</span> <span class="va">metabolomic</span><span class="op">[</span>, <span class="va">rng2</span><span class="op">]</span>, ratio <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme</span><span class="op">(</span>text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/correlation%201-1.png" style="display: block; margin: auto;"></p>
<div style="text-align: justify">
<p>By default the function will keep every features, you can specify a cutoff of correlation factor, here we will keep only the features were at least one as a correlation factor &gt; 0.7 and with a p.adj &lt; 0.05. The function will only return features checking those two criteria, for now you can’t chose between corrected and non corrected p values.<br>
You can specify two different matrices, for example the taxa table and the features to find correlations between these, just make sure you’re data are in the correct order because the function assumes that the samples are matching between the two matrices.</p>
</div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_corr_heatmap.html">plot_corr_heatmap</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">metabolomic</span><span class="op">[</span>, <span class="va">rng</span><span class="op">]</span>, Y <span class="op">=</span> <span class="va">metabolomic</span><span class="op">[</span>, <span class="va">rng2</span><span class="op">]</span>, cutoff <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/correlation%202-1.png" style="display: block; margin: auto;"></p>
<p>Finally, you can also chose to cluster the features both for x and y axis.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_corr_heatmap.html">plot_corr_heatmap</a></span><span class="op">(</span>X <span class="op">=</span> <span class="va">metabolomic</span><span class="op">[</span>, <span class="va">rng</span><span class="op">]</span>, Y <span class="op">=</span> <span class="va">metabolomic</span><span class="op">[</span>, <span class="va">rng2</span><span class="op">]</span>, cutoff <span class="op">=</span> <span class="fl">0.4</span>,</span>
<span>    cluster <span class="op">=</span> <span class="cn">T</span>, ratio <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/correlation%203-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="volcano-and-other-plots-to-represent-enrichment">Volcano and other plots to represent enrichment<a class="anchor" aria-label="anchor" href="#volcano-and-other-plots-to-represent-enrichment"></a></h3>
<div style="text-align: justify">
<p>Finally, we finish with enrichment/differential abundance analysis for non phyloseq objects. This is a big part of biostat analysis, find the features/taxa differently abundant between two groups of patient/mice. There are a lot of discussion about how to normalize data in the microbiomics, this might be the same case for metabolomics and metagenomics. Here the function use a classic Mann-Whitney or Student test with a FDR correction. So this is the most basic approach. You can chose to normalize the data before giving it to the function, this will be on you to chose which approach is better.</p>
<div class="section level4">
<h4 id="volcano-plots">Volcano plots<a class="anchor" aria-label="anchor" href="#volcano-plots"></a></h4>
<p>We already have a function producing volcano plots based on the ALDeX2 package but not for non phyloseq objects.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fold</span> <span class="op">=</span> <span class="fu"><a href="reference/fold_change.html">fold_change</a></span><span class="op">(</span>clinical_data <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="va">metabolomic</span>, cores <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_volcano.html">plot_volcano</a></span><span class="op">(</span><span class="va">fold</span><span class="op">$</span><span class="va">birth_type</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: ggrepel: 6 unlabeled data points (too many overlaps). Consider</span></span>
<span><span class="co">#&gt; increasing max.overlaps</span></span></code></pre></div>
<p><img src="ImmuMicrobiome_book_files/figure-gfm/fold%20change-1.png" style="display: block; margin: auto;"></p>
</div>
<div class="section level4">
<h4 id="ternary-plots">Ternary plots<a class="anchor" aria-label="anchor" href="#ternary-plots"></a></h4>
<p>The function is not yet created but this following code will create a ternary plot. Bare in mind that this function will accept only three groups, no more no less. For two groups use the volcano function, for more than three groups well you nothing exists yet.</p>
</div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="iga-seq-analysis-pipeline">IgA seq analysis pipeline<a class="anchor" aria-label="anchor" href="#iga-seq-analysis-pipeline"></a></h1>
<div style="text-align: justify">
<p>You’ve performed all your sortings and sequencing, you now have three samples coming from a single individual.<br>
We created a pipeline analysis where you will use the <strong>neg1</strong> (9/10 of the neg fraction) and <strong>neg2</strong> (1/10 of the neg fraction) dispersion (centered and reduced) to create a normal dispersion, using a Z approach we will have a Z score for the based on the standard deviation of .</p>
<p>Here we can see what the analysis will look like:</p>
<div class="figure" style="text-align: center">
<img src="../Documentation_ImmuMicrobiome/Z%20test.png" alt="Scheme of Z test approach" width="2700"><p class="caption">
Scheme of Z test approach
</p>
</div>
<ul><li><p>The technical dispersion is assessed using <code>neg1/neg2</code>for the log2 ratio and the <code>neg1*neg2</code>abundance for log10 (black dot)</p></li>
<li><p>The biological dispersion is assessed <code>pos/neg1</code> for the log2 ratio and the <code>pos*neg1</code> abundance for log10 (orange dot)</p></li>
</ul><p>We will then take windows of X ASV (for example 20) to create n Gaussian curve and n …<br>
The pipeline is based on three main functions that will call for other functions : seq_table, slide_z and collapse_IgAseq.</p>
</div>
<div class="section level2">
<h2 id="make-the-seq_table-list">Make the seq_table list<a class="anchor" aria-label="anchor" href="#make-the-seq_table-list"></a></h2>
<div style="text-align: justify">
<p>First the seq_table function will take your phyloseq object and transform it to a list of data frames, one data frame for each samples coming from a single individual. For this function you will need to give physeq, sample_name corresponding to the name identifying the individual from which the sorted samples came, sorting_names the column where we find the samples such as : “sample1_pos”, “sample1_neg1”, “sample1_neg2”. Then the cols_to_keep that need to stay for now on “all”, it will collapse your sample_data in one column to allow you get it back later on.</p>
<p>The function will tell you if there is some samples are alone, if you have duplicated samples you need to sort them out or the rest of the pipeline will block. You can take a look at the architecture of the new object, you will find the ASV sequence as rownames, the taxonomy collapsed with “#” separator, the three samples having their own column and the sample_data collapse using also “#” separator. The function will only take the ASV that are present in the samples.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"igaseq"</span><span class="op">)</span></span>
<span><span class="va">igaseq</span> <span class="op">=</span> <span class="fu">transform_sample_counts</span><span class="op">(</span><span class="va">igaseq</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># sample_names(igaseq)= sample_data(igaseq)$sort_type</span></span>
<span></span>
<span><span class="va">seq.tab</span> <span class="op">=</span> <span class="fu"><a href="reference/seq_table.html">seq_table</a></span><span class="op">(</span><span class="va">igaseq</span>, sample_name <span class="op">=</span> <span class="st">"sample_origin"</span>, sorting_names <span class="op">=</span> <span class="st">"sample_sort"</span>,</span>
<span>    cols_to_keep <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span></span>
<span><span class="co">#&gt; sample MO308 is alone </span></span>
<span><span class="co">#&gt; One sample belonging to C1636 has no reads</span></span>
<span><span class="co">#&gt; sample C1232 is alone </span></span>
<span><span class="co">#&gt; sample C1283 is alone </span></span>
<span><span class="co">#&gt; sample T2648 is alone</span></span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">seq.tab</span><span class="op">$</span><span class="va">C1101</span><span class="op">)</span>, rownames <span class="op">=</span> <span class="cn">F</span>, booktabs <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<table class="table"><colgroup><col width="67%"><col width="17%"><col width="1%"><col width="1%"><col width="1%"><col width="1%"><col width="8%"></colgroup><thead><tr class="header"><th align="left">FALSE</th>
<th align="left">taxonomy</th>
<th align="right">C1101_pos</th>
<th align="right">C1101_neg2</th>
<th align="right">C1101_neg1</th>
<th align="left">sample_id</th>
<th align="left">new</th>
</tr></thead><tbody><tr class="odd"><td align="left">AGTGGGGAATATTGGGCAATGGGGGAAACCCTGACCCAGCAACGCCGCGTGAAGGAAGAAGGCCTTCGGGTTGTAAACTTCTTTTACCAGGGACGAAGGACGTGACGGTACCTGGAGAAAAAGCAACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTTGCAAGCGTTGTCCGGATTTACTGGGTGTAAAGGGCGTGTAGGCGGAGATGCAAGTTGGGAGTGAAATCCATGGGCTCAACCCATGAACTGCTCTCAAAACTGTATCCCTTGAGTATCGGAGAGGCAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTGCTGGACGACAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGGATTA</td>
<td align="left">Bacteria#Firmicutes#Clostridia#Oscillospirales#Oscillospiraceae#UCG-005#NA#UCG-005_372</td>
<td align="right">0.0000000</td>
<td align="right">0.0011692</td>
<td align="right">0.0008873</td>
<td align="left">C1101</td>
<td align="left">C1101_pos#C1101#FAFR101#M#breastfed#Vaginal#infant</td>
</tr><tr class="even"><td align="left">AGTGGGGGATATTGCACAATGGGGGAAACCCTGATGCAGCAACGCCGCGTGAGGGAAGAAGGTTTTCGGATTGTAAACCTCTGTTCTTAGTGACGATAATGACGGTAGCTAAGGAGAAAGCTCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGAGCGAGCGTTGTCCGGATTTACTGGGTGTAAAGGGTGCGTAGGCGGCGAGGCAAGTCAGGCGTGAAATCTATGGGCTTAACCCATAAACTGCGCTTGAAACTGTCTTGCTTGAGTGAAGTAGAGGTAGGCGGAATTCCCGGTGTAGCGGTGAAATGCGTAGAGATCGGGAGGAACACCAGTGGCGAAGGCGGCCTACTGGGCTTTAACTGACGCTGAAGCACGAAAGCATGGGTAGCAAACAGGATTA</td>
<td align="left">Bacteria#Firmicutes#Clostridia#Oscillospirales#Ruminococcaceae#Incertae Sedis#NA#Incertae Sedis_437</td>
<td align="right">0.0000000</td>
<td align="right">0.0017987</td>
<td align="right">0.0023662</td>
<td align="left">C1101</td>
<td align="left">C1101_pos#C1101#FAFR101#M#breastfed#Vaginal#infant</td>
</tr><tr class="odd"><td align="left">AGTCGGGAATATTGCGCAATGGAGGAAACTCTGACGCAGTGACGCCGCGTATGGGAAGAAGGTTTTCGGATTGTAAACCATTTTAGACAAGGAAGAAACAAGACAGTACTTGTAGAATAAGCTCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGAGCAAGCGTTATCCGGATTTATTGGGTGTAAAGGGTGCGTAGACGGGAAGGTAAGTTAGTTGTGAAATCCCTCGGCTCAACTGAGGAACTGCGACTAAAACTGCTTTTCTTGAGTGCTGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGACTTTCTGGACAGTAACTGACGTTGAGGCACGAAAGTGTGGGGAGCAAACAGGATTA</td>
<td align="left">Bacteria#Firmicutes#Clostridia#Clostridia UCG-014#NA#NA#NA#NA_486</td>
<td align="right">0.0012804</td>
<td align="right">0.0009893</td>
<td align="right">0.0000000</td>
<td align="left">C1101</td>
<td align="left">C1101_pos#C1101#FAFR101#M#breastfed#Vaginal#infant</td>
</tr><tr class="even"><td align="left">AGTGGGGAATATTGCACAATGGGGGAAACCCTGATGCAGCAACGCCGCGTGAGTGAAGAAGTATTTCGGTATGTAAAGCTCTATCAGCAGGGAAGAAAATGACGGTACCTGACTAAGAAGCACCGGCTAAATACGTGCCAGCAGCCGCGGTAATACGTATGGTGCAAGCGTTATCCGGATTTACTGGGTGTAAAGGGAGCGCAGGCGGTCTGGCAAGTCTGATGTGAAAGCCCGGGGCTCAACCCCGGGACTGCATTGGAAACTGTCAGACTAGAGTGTCGGAGAGGTAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACACCAGTGGCGAAGGCGGCTTACTGGACGATAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTA</td>
<td align="left">Bacteria#Firmicutes#Clostridia#Lachnospirales#Lachnospiraceae#NA#NA#NA_635</td>
<td align="right">0.0000000</td>
<td align="right">0.0045867</td>
<td align="right">0.0040422</td>
<td align="left">C1101</td>
<td align="left">C1101_pos#C1101#FAFR101#M#breastfed#Vaginal#infant</td>
</tr><tr class="odd"><td align="left">AGTGGGGAATATTGCACAATGGAGGAAACTCTGATGCAGCGATGCCGCGTGAGGGAAGAAGGTTTTAGGATTGTAAACCTCTGTCTTCAGGGACGAAAAAAAAGACGGTACCTGAGGAGGAAGCTCCGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGGAGCGAGCGTTGTCCGGAATTACTGGGTGTAAAGGGAGCGTAGGCGGGATCGCAAGTCAGATGTGAAAACTATGGGCTTAACCCATAAACTGCATTTGAAACTGTGGTTCTTGAGTGAAGTAGAGGTAAGCGGAATTCCTAGTGTAGCGGTGAAATGCGTAGATATTAGGAGGAACATCAGTGGCGAAGGCGGCTTACTGGGCTTTAACTGACGCTGAGGCTCGAAAGCGTGGGGAGCAAACAGGATTA</td>
<td align="left">Bacteria#Firmicutes#Clostridia#Oscillospirales#Ruminococcaceae#Ruminococcus#bicirculans#Ruminococcus_785</td>
<td align="right">0.0021668</td>
<td align="right">0.0041371</td>
<td align="right">0.0020704</td>
<td align="left">C1101</td>
<td align="left">C1101_pos#C1101#FAFR101#M#breastfed#Vaginal#infant</td>
</tr><tr class="even"><td align="left">AGTGGGGAATCTTCCGCAATGGGCGAAAGCCTGACGGAGCAACGCCGCGTGAGTGATGACGGCCTTCGGGTTGTAAAGCTCTGTGATCGGGGACGAACGGTCTGTAAGCTAATATCTTATGGAAGTGACGGTACCCGAATAGCAAGCCACGGCTAACTACGTGCCAGCAGCCGCGGTAATACGTAGGTGGCAAGCGTTGTCCGGAATTATTGGGCGTAAAGCGCGCGCAGGCGGCTTTCTAAGTCCATCTTAAAAGTGCGGGGCTTAACCCCGTGATGGGATGGAAACTGGAAAGCTGGAGTATCGGAGAGGAAAGTGGAATTCCTAGTGTAGCGGTGAAATGCGTAGAGATTAGGAAGAACACCGGTGGCGAAGGCGACTTTCTGGACGACAACTGACGCTGAGGCGCGAAAGCGTGGGGAGCAAACAGGATTA</td>
<td align="left">Bacteria#Firmicutes#Negativicutes#Veillonellales-Selenomonadales#Veillonellaceae#Dialister#NA#Dialister_1138</td>
<td align="right">0.0000000</td>
<td align="right">0.0046767</td>
<td align="right">0.0054225</td>
<td align="left">C1101</td>
<td align="left">C1101_pos#C1101#FAFR101#M#breastfed#Vaginal#infant</td>
</tr></tbody></table><p>The colnames will have the sample_names as given in the otu_table(), check that they correspond to your pos, neg1 and neg2.</p>
<p>Run the main function Now we will run the main function : slide_z. This function will take you seq_table object and run the Z function for each samples. If you are running this function for the first time use plot=T, if you already made the plots let it as FALSE it will make the loop a lot faster.</p>
</div>
</div>
<div class="section level2">
<h2 id="deal-with-the-zero-values">Deal with the zero values<a class="anchor" aria-label="anchor" href="#deal-with-the-zero-values"></a></h2>
<div style="text-align: justify">
<p>In this approach we will use log2 ratio and log10 of abundances. As you know log2(0/x) or log2(x/0) can’t be performed, so we need a way to deal with the zeros. We came up with two approach :</p>
<ul><li><p>remove all ASV with a zero value</p></li>
<li><p>replace 0 by a random number between 0 and the min value found in one of the three samples</p></li>
<li><p>replace 0 by the minimum count in each samples, this probably the worst thing to do</p></li>
</ul><p>In sample with few ASV, i.e meconiums, I strongly recommend using the random_generation while in adults samples the no_zero approach is performing better. For more complex samples the decision is up to you, you can use the function neg_dipsersion to visualize the two outcomes.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/neg_dispersion.html">neg_dispersion</a></span><span class="op">(</span><span class="va">seq.tab</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, positive_sorted_sample <span class="op">=</span> <span class="st">"pos"</span>, negative_sorted_sample <span class="op">=</span> <span class="st">"neg1"</span>,</span>
<span>    second_negative_sample <span class="op">=</span> <span class="st">"neg2"</span>, type <span class="op">=</span> <span class="st">"superposed"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="reference/neg_dispersion.html">neg_dispersion</a></span><span class="op">(</span><span class="va">seq.tab</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, positive_sorted_sample <span class="op">=</span> <span class="st">"pos"</span>, negative_sorted_sample <span class="op">=</span> <span class="st">"neg1"</span>,</span>
<span>    second_negative_sample <span class="op">=</span> <span class="st">"neg2"</span>, type <span class="op">=</span> <span class="st">"facet all three"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-16-1.png" alt="Verify the dispersion of the negative fractions"><img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-16-2.png" alt="Verify the dispersion of the negative fractions"><p class="caption">
Verify the dispersion of the negative fractions
</p>
</div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run the following if you want every samples, make sure to change the number</span></span>
<span><span class="co"># of cores pdf('test.pdf', width = 10, height = 7) mclapply(seq,</span></span>
<span><span class="co"># neg_dispersion, mc.cores = 6, type='superposed') dev.off()</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="run-the-slide_z-function">Run the slide_z function<a class="anchor" aria-label="anchor" href="#run-the-slide_z-function"></a></h2>
<div style="text-align: justify">
<p>The first step will be to create log2 ratios and log10 abundance (log10 abundance of <strong>pos</strong> * <strong>neg1</strong> for example) for each ASV between the <strong>pos</strong> and the <strong>neg1</strong> and between the <strong><em>neg1</em></strong> and <strong>neg2</strong> . This will be done by the function <code>log_ratio</code> called by the <code>slide_z</code> function.<br>
The function will also create an ellipse of confidence interval of your choosing, default being <code>confidence_interval= c(0.95,0.99, 0.999)</code>, this is done by the function <code>ellipse_me</code> also called by the <code>slide_z</code> function.</p>
<p>The output will be a list of S4 objects containing the following slots :</p>
<ul><li><p><em>ig_seq_all</em> : containing all the samples and all the ASV</p></li>
<li><p><em>ig_up</em> : containing all the samples and all the ASV that significantly enriched in the <strong>IgA positive fraction</strong></p></li>
<li><p><em>ig_down</em> : containing all the samples and all the ASV that are significantly enriched in the <strong>IgA negative fraction</strong></p></li>
</ul><p>In each slot you will find the following columns :</p>
<ul><li><p><em>taxonomy</em></p></li>
<li><p><em>sample_id</em></p></li>
<li><p><em>new</em> : the sample_data collapsed</p></li>
<li><p><em>pos</em> : positive fraction abundance</p></li>
<li><p><em>neg1</em> : negative (9/10) fraction abundance</p></li>
<li><p><em>neg2</em> : negative (1/10) fraction abundance</p></li>
<li><p><em>log10_abundance</em> = log10(pos * neg1)</p></li>
<li><p><em>log2_ratio</em> = log2(pos / neg1)</p></li>
<li><p><em>log10_neg_abundance</em> = log10(neg1 * neg2)</p></li>
<li><p><em>log2_neg_ratio</em> = log2(neg1 / neg2)</p></li>
<li><p><em>taxa</em> = tax_table collapsed</p></li>
<li><p><em>SlideNorm</em> = the normalized dispersion for each ASV</p></li>
<li><p><em>score</em> = IgAseq score</p></li>
<li><p><em>ellipse_level</em> = the level of confidence</p></li>
</ul><div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">IgA_seq</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">seq.tab</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co"># print(i)</span></span>
<span>    <span class="va">IgA_seq</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="reference/slide_z.html">slide_z</a></span><span class="op">(</span><span class="va">seq.tab</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, positive_sorted_sample <span class="op">=</span> <span class="st">"pos"</span>, negative_sorted_sample <span class="op">=</span> <span class="st">"neg1"</span>,</span>
<span>        second_negative_sample <span class="op">=</span> <span class="st">"neg2"</span>, deltaX <span class="op">=</span> <span class="fl">30</span>, slide_version <span class="op">=</span> <span class="st">"slide_z_modern"</span>,</span>
<span>        alpha <span class="op">=</span> <span class="fl">0.05</span>, plot <span class="op">=</span> <span class="cn">F</span>, zero_treatment <span class="op">=</span> <span class="st">"random generation"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Just for the demo we will make the plots with <code>plot= T</code>. This will plot .png and .hmtl <code>plotly</code>files. This will make the loop slower but it will allow you to verify the distribution of your samples and potentially enable you to remove outliers.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">=</span> <span class="fu"><a href="reference/slide_z.html">slide_z</a></span><span class="op">(</span><span class="va">seq.tab</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, positive_sorted_sample <span class="op">=</span> <span class="st">"pos"</span>, negative_sorted_sample <span class="op">=</span> <span class="st">"neg1"</span>,</span>
<span>    second_negative_sample <span class="op">=</span> <span class="st">"neg2"</span>, deltaX <span class="op">=</span> <span class="fl">30</span>, slide_version <span class="op">=</span> <span class="st">"slide_z_modern"</span>,</span>
<span>    alpha <span class="op">=</span> <span class="fl">0.05</span>, plot <span class="op">=</span> <span class="cn">T</span>, zero_treatment <span class="op">=</span> <span class="st">"random generation"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="Igaseq%20plots/png/MO101_random%20generation_.png" alt="Example of the plots output, a .png and a .html are saved" width="2100"><p class="caption">
Example of the plots output, a .png and a .html are saved
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="collapse-the-list-of-igaseq-into-a-single-s4-object">Collapse the list of IgAseq into a single S4 object<a class="anchor" aria-label="anchor" href="#collapse-the-list-of-igaseq-into-a-single-s4-object"></a></h2>
<div style="text-align: justify">
<p>We now need to collapse the list of IgA_seq objects into a single one, just use <code>collapse_IgAseq</code> and separate the columns that were pasted together.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">IgA_seq</span> <span class="op">=</span> <span class="fu"><a href="reference/collapse_IgAseq.html">collapse_IgAseq</a></span><span class="op">(</span><span class="va">IgA_seq</span><span class="op">)</span></span>
<span><span class="co"># DT:: datatable(IgA_seq@ig_seq_all, rownames = F) View(IgA_seq)</span></span>
<span></span>
<span></span>
<span><span class="va">IgA_all</span> <span class="op">=</span> <span class="va">IgA_seq</span><span class="op">@</span><span class="va">ig_seq_all</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">separate</span><span class="op">(</span><span class="va">taxonomy</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Reign"</span>, <span class="st">"Phylum"</span>, <span class="st">"Order"</span>, <span class="st">"Class"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>,</span>
<span>        <span class="st">"Species"</span>, <span class="st">"ASV"</span>, <span class="st">"rest"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">separate</span><span class="op">(</span>col <span class="op">=</span> <span class="va">new</span>, into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">sample_data</span><span class="op">(</span><span class="va">igaseq</span><span class="op">)</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"#"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">IgA_all</span><span class="op">$</span><span class="va">alpha</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">IgA_all</span><span class="op">$</span><span class="va">score</span> <span class="op">&gt;</span> <span class="fl">1.96</span>, <span class="st">"positively significative"</span>, <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">IgA_all</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;</span></span>
<span>    <span class="op">-</span><span class="fl">1.96</span>, <span class="st">"negatively significative"</span>, <span class="st">"not significative"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">IgA_all</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 4222   29</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="analyse-and-plot-igaseq">Analyse and plot IgASeq<a class="anchor" aria-label="anchor" href="#analyse-and-plot-igaseq"></a></h2>
<div class="section level3">
<h3 id="plot-like-gordon-igaseq">Plot like Gordon IgASeq<a class="anchor" aria-label="anchor" href="#plot-like-gordon-igaseq"></a></h3>
<div style="text-align: justify">
<p>For the example we will plot the IgAseq data like Gordon’s paper. We first make a wilcoxon test to decipher which genera are significantly different from zero. Then we plot as balloon plot using ggplot.</p>
</div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/rstatix/" class="external-link">rstatix</a></span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">=</span> <span class="va">IgA_all</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Genus</span>, <span class="va">donor</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="va">Genus</span> <span class="op">!=</span> <span class="st">"NA"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/wilcox_test.html" class="external-link">wilcox_test</a></span><span class="op">(</span><span class="va">score</span> <span class="op">~</span> <span class="fl">1</span>, mu <span class="op">=</span> <span class="fl">0</span>, alternative <span class="op">=</span> <span class="st">"two.sided"</span>, detailed <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tmp2</span> <span class="op">=</span> <span class="va">IgA_all</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">Genus</span>, <span class="va">donor</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="va">Genus</span> <span class="op">!=</span> <span class="st">"NA"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Phylum</span>, <span class="va">Genus</span>, <span class="va">score</span>, <span class="va">sample_origin</span>, <span class="va">donor</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>mean_score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">left_join</span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tmp2</span><span class="op">$</span><span class="va">mean_score</span><span class="op">[</span><span class="va">tmp2</span><span class="op">$</span><span class="va">mean_score</span> <span class="op">&lt;=</span> <span class="op">(</span><span class="op">-</span><span class="fl">5</span><span class="op">)</span><span class="op">]</span> <span class="op">=</span> <span class="op">(</span><span class="op">-</span><span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">alpha</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">tmp2</span><span class="op">$</span><span class="va">p</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">tmp2</span><span class="op">$</span><span class="va">p</span><span class="op">)</span>, <span class="fl">0.5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tmp2</span> <span class="op">=</span> <span class="va">tmp2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">score</span>, <span class="va">mean_score</span>, <span class="va">Phylum</span>, <span class="va">Genus</span>, <span class="va">donor</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">tmp2</span><span class="op">$</span><span class="va">p</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">tmp2</span><span class="op">$</span><span class="va">p</span><span class="op">)</span>, <span class="fl">0</span><span class="op">)</span>, alpha2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">tmp2</span><span class="op">$</span><span class="va">mean_score</span> <span class="op">&lt;</span></span>
<span>        <span class="fl">0</span>, <span class="op">-</span><span class="va">alpha</span>, <span class="va">alpha</span><span class="op">)</span>, mean_score2 <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">mean_score</span><span class="op">)</span><span class="op">)</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,</span>
<span>        <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="va">tmp2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="fl">0</span>, <span class="va">Genus</span>, fill <span class="op">=</span> <span class="va">alpha2</span>, size <span class="op">=</span> <span class="va">mean_score2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_point</span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_fill_gradient2</span><span class="op">(</span>low <span class="op">=</span> <span class="st">"olivedrab4"</span>, mid <span class="op">=</span> <span class="st">"white"</span>, high <span class="op">=</span> <span class="st">"brown"</span>, midpoint <span class="op">=</span> <span class="fl">0</span>,</span>
<span>        guide <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_size_continuous</span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>,</span>
<span>    <span class="op">-</span><span class="fl">4</span>, <span class="op">-</span><span class="fl">3</span>, <span class="fl">0</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span><span class="op">)</span>, range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">guides</span><span class="op">(</span>size <span class="op">=</span> <span class="fu">guide_legend</span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"darkgreen"</span>,</span>
<span>    <span class="st">"white"</span>, <span class="st">"brown"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">7</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">4</span>, <span class="fl">3</span>, <span class="fl">1</span>, <span class="fl">3</span>, <span class="fl">4</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span>, direction <span class="op">=</span> <span class="st">"horizontal"</span>,</span>
<span>    title.position <span class="op">=</span> <span class="st">"top"</span>, label.position <span class="op">=</span> <span class="st">"bottom"</span>, label.hjust <span class="op">=</span> <span class="fl">0.5</span>, label.vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">labs</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="st">"Age"</span>, size <span class="op">=</span> <span class="st">"IgAseq median score"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">facet_grid</span><span class="op">(</span><span class="va">Phylum</span> <span class="op">~</span></span>
<span>    <span class="va">donor</span>, scales <span class="op">=</span> <span class="st">"free"</span>, space <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>    face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, axis.ticks.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    strip.text.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, axis.title.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    legend.position <span class="op">=</span> <span class="st">"bottom"</span>, legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,</span>
<span>    legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, legend.box <span class="op">=</span> <span class="st">"vertical"</span>,</span>
<span>    legend.box.background <span class="op">=</span> <span class="fu">element_rect</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>, strip.background <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    strip.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, panel.grid.major.y <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>linetype <span class="op">=</span> <span class="fl">2</span>,</span>
<span>        size <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `size` argument of `element_line()` is deprecated as of ggplot2 3.4.0.</span></span>
<span><span class="co">#&gt; ℹ Please use the `linewidth` argument instead.</span></span>
<span><span class="co">#&gt; This warning is displayed once every 8 hours.</span></span>
<span><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span><span class="co">#&gt; generated.</span></span>
<span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="va">tmp2</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">score</span>, <span class="va">Genus</span>, fill <span class="op">=</span> <span class="va">alpha2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_boxplot</span><span class="op">(</span>outlier.size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_fill_gradient2</span><span class="op">(</span>low <span class="op">=</span> <span class="st">"olivedrab4"</span>, mid <span class="op">=</span> <span class="st">"white"</span>, high <span class="op">=</span> <span class="st">"brown"</span>, midpoint <span class="op">=</span> <span class="fl">0</span>,</span>
<span>        guide <span class="op">=</span> <span class="cn">F</span><span class="op">)</span> <span class="op">+</span> <span class="fu">facet_grid</span><span class="op">(</span><span class="va">Phylum</span> <span class="op">~</span> <span class="va">donor</span>, scales <span class="op">=</span> <span class="st">"free_y"</span>, space <span class="op">=</span> <span class="st">"free"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>axis.text.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, strip.text.y <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">0</span>, size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>        face <span class="op">=</span> <span class="st">"bold"</span>, hjust <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>, axis.ticks <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, axis.title.x <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    axis.title.y <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span>, legend.title <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>        face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, legend.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, legend.box <span class="op">=</span> <span class="st">"vertical"</span>,</span>
<span>    legend.box.background <span class="op">=</span> <span class="fu">element_rect</span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span>, strip.background <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>,</span>
<span>    strip.text <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>, panel.grid.major.y <span class="op">=</span> <span class="fu">element_line</span><span class="op">(</span>linetype <span class="op">=</span> <span class="fl">2</span>,</span>
<span>        size <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">ggpubr</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, nrow <span class="op">=</span> <span class="fl">1</span>, widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, common.legend <span class="op">=</span> <span class="cn">T</span>, legend <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `guide` argument in `scale_*()` cannot be `FALSE`. This was deprecated in</span></span>
<span><span class="co">#&gt; ggplot2 3.3.4.</span></span>
<span><span class="co">#&gt; ℹ Please use "none" instead.</span></span>
<span><span class="co">#&gt; This warning is displayed once every 8 hours.</span></span>
<span><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span><span class="co">#&gt; generated.</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-22-1.png" alt="Bubble plot of IgASeq score"><p class="caption">
Bubble plot of IgASeq score
</p>
</div>
</div>
<div class="section level3">
<h3 id="reduction-of-dimensions-on-igaseq">Reduction of dimensions on IgASeq<a class="anchor" aria-label="anchor" href="#reduction-of-dimensions-on-igaseq"></a></h3>
<p>IgASeq is a particular type of data. Indeed, it is not compositional anymore and only values outside [-1.96 ; +1.96] are considered as statistically significant. What should be done for the taxa that are comprised in this interval ? Good question.</p>
<p>No matter what, you cannot use classical distance algorithm such as Bray-Curtis or Jaccard : those algorithm doesn’t handle negatie values. We only have Euclidean and Canberra algorithm that can handle our new type of data. It seems that Canberra handles well 0 centered values. So we encourage the use of Canberra distance. The reduction approach is up to you.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">=</span> <span class="va">IgA_all</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">sample_origin</span>, <span class="va">ASV</span>, <span class="va">score</span>, <span class="va">donor</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">summarise</span><span class="op">(</span>score <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">score</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">pivot_wider</span><span class="op">(</span>values_from <span class="op">=</span> <span class="va">score</span>, names_from <span class="op">=</span> <span class="va">ASV</span>, values_fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="va">as.data.frame</span></span>
<span></span>
<span><span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="reference/plot_reduction.html">plot_reduction</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"PCA"</span>, clinical_data <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, dist <span class="op">=</span> <span class="st">"canberra"</span>, group <span class="op">=</span> <span class="st">"donor"</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"pure"</span>, draw <span class="op">=</span> <span class="st">"polygon"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="reference/plot_reduction.html">plot_reduction</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"PCoA"</span>, clinical_data <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, dist <span class="op">=</span> <span class="st">"canberra"</span>, group <span class="op">=</span> <span class="st">"donor"</span>,</span>
<span>        type <span class="op">=</span> <span class="st">"pure"</span>, draw <span class="op">=</span> <span class="st">"polygon"</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-23-1.png" alt="Reduction of IgASeq score"><img src="ImmuMicrobiome_book_files/figure-gfm/unnamed-chunk-23-2.png" alt="Reduction of IgASeq score"><p class="caption">
Reduction of IgASeq score
</p>
</div>
<p>As you can see PCA can be strongly impacted by extreme values.</p>
</div>
</div>
</div>


  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

