<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyse microbiome, metabolome and IgASeq. • ImmuMicrobiome</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Analyse microbiome, metabolome and IgASeq.">
<meta property="og:description" content="This package is created to analyse microbiome data with phyloseq objects. Plot function are based on publication of Villette et al. and propose a core plotting outcome. This package also implement wrapper functions for dada2 and a IgASeq pipeline.Function in this package allows to : analyse alha-diversity, beta-diversity, correlations, differential abundance, etc. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ImmuMicrobiome</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/ImmuMicrobiome_vignette.html">ImmuMicrobiome_vignette</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="immumicrobiome">ImmuMicrobiome<a class="anchor" aria-label="anchor" href="#immumicrobiome"></a>
</h1></div>
<p>This package is based on the code of Dr. Villette and Dr. Larsen. The package is written and maintained by Dr. Villette. This package is meant to facilitate microbiome exploration and ensuring nice plotting.<br>
This package covers :<br></p>
<ul>
<li><p>The DADA2 pipeline with wrapper functions that ease the processing of multiple projects</p></li>
<li><p>Some plotting functions for beta diversity, heatmap and differential abundance analysis using directly a phyloseq object</p></li>
<li><p>A pipeline for <strong>IgASeq</strong> analysis</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="dada2-pipeline">DADA2 pipeline<a class="anchor" aria-label="anchor" href="#dada2-pipeline"></a>
</h2>
<div style="text-align: justify">
<p>I strongly encourage you to read the paper from Callahan and also the <a href="https://benjjneb.github.io/dada2/tutorial.html" class="external-link">tutorial</a>. DADA2 (and other pipeline) introduced the concept of Exact Sequence Variant which are very different from the OTU approach.<br>
Quickly, ESV (ASV for DADA2) considers that every reads with a single base difference is different from another by definition. In this approach the authors came with the idea that until proven otherwise a read is different from another even when only a single base is different.<br>
To decipher if the read is truly different or not, the pipeline will use the Phred score associated to each base to determine if the base read here is different between a read A and B because of:</p>
<ul>
<li><p>A biological difference, for example to different strains</p></li>
<li><p>A technical difference, a misreading from the sequencer</p></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="nice-plotting">Nice plotting.<a class="anchor" aria-label="anchor" href="#nice-plotting"></a>
</h2>
<div style="text-align: justify">
<p>Microbiomics is a very nice playground for biostatistic and data modelisation. Based on the work done in my lab, I’ve created a bunch of wrapper functions and new functions to explore microbiomics. A part of the functions are working directly on <a href="https://joey711.github.io/phyloseq/" class="external-link">phyloseq</a> objects and the others accept directly data.frames.</p>
</div>
</div>
<div class="section level2">
<h2 id="igaseq-analysis">IgASeq analysis.<a class="anchor" aria-label="anchor" href="#igaseq-analysis"></a>
</h2>
<div style="text-align: justify">
<p>IgA and more importantly secretory IgA gain a growing interest as it is the main immunological microbial regulator. This immunoglobulin has the particular ability to control both pathogens and commensals, with a negative action on pathogens (pathogen removal) and a positive action on commensals (ecological niche formation).<br>
In this package we describe a new method to produce the samples and also to make the analysis after the sequencing. This method is actually published in […].</p>
</div>
</div>
<div class="section level2">
<h2 id="machine-learning">Machine learning<a class="anchor" aria-label="anchor" href="#machine-learning"></a>
</h2>
<div style="text-align: justify">
<p>While we are not expert on machine learning and statistical models, more and more authors use them to predict or classify samples according to variables and features, we decided to support machine learning steps in this package. In this package we made wrapper functions to ease the use of models.</p>
<p>This part is not yet furnished and will available in future updates.</p>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ImmuMicrobiome</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>The package maintainer <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
