<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ ImmuMicrobiome</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ImmuMicrobiome</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/ImmuMicrobiome_vignette.html">ImmuMicrobiome_vignette</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="alpha_diversity.html">alpha_diversity()</a></code> </p>
        </td>
        <td><p>Title Wrapper function to analyse alpha diversity.</p></td>
      </tr><tr><td>
          <p><code><a href="beta_dispersion.html">beta_dispersion()</a></code> </p>
        </td>
        <td><p>Make an unconstrained analysis of beta diversity.</p></td>
      </tr><tr><td>
          <p><code><a href="beta_diversity.html">beta_diversity()</a></code> </p>
        </td>
        <td><p>Plot beta-diversity with numerous aspect.</p></td>
      </tr><tr><td>
          <p><code><a href="calculate_variance.html">calculate_variance()</a></code> </p>
        </td>
        <td><p>Calculate the variance for each factors.</p></td>
      </tr><tr><td>
          <p><code><a href="check_length.html">check_length()</a></code> </p>
        </td>
        <td><p>Check the length of reads in fastq.</p></td>
      </tr><tr><td>
          <p><code><a href="circular_variance_plot.html">circular_variance_plot()</a></code> </p>
        </td>
        <td><p>Make circular plots for variance analysis.</p></td>
      </tr><tr><td>
          <p><code><a href="clean_sra_metadata.html">clean_sra_metadata()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="collapse_IgAseq.html">collapse_IgAseq()</a></code> </p>
        </td>
        <td><p>Collapse IgASeq</p></td>
      </tr><tr><td>
          <p><code><a href="constrained_beta_dispersion.html">constrained_beta_dispersion()</a></code> </p>
        </td>
        <td><p>Make a constrained analysis of beta diversity.</p></td>
      </tr><tr><td>
          <p><code><a href="cross_validate_data.html">cross_validate_data()</a></code> </p>
        </td>
        <td><p>Cross validate the training data.</p></td>
      </tr><tr><td>
          <p><code><a href="dereplicate.html">dereplicate()</a></code> </p>
        </td>
        <td><p>Dereplicate sequences in a dada2 manner</p></td>
      </tr><tr><td>
          <p><code><a href="differential_abundance.html">differential_abundance()</a></code> </p>
        </td>
        <td><p>Title
Find differential abundance within a dataset.</p></td>
      </tr><tr><td>
          <p><code><a href="ellipse_me.html">ellipse_me()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="error_check.html">error_check()</a></code> </p>
        </td>
        <td><p>Error estimation according to dada2</p></td>
      </tr><tr><td>
          <p><code><a href="filter_fastq.html">filter_fastq()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="filt_list.html">filt_list()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="find_primer.html">find_primer()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="FoldChangeClass-class.html">FoldChangeClass-class</a></code> <code><a href="FoldChangeClass-class.html">fold</a></code> </p>
        </td>
        <td><p>Create a S4 object for the IgASeq data</p></td>
      </tr><tr><td>
          <p><code><a href="fold_change.html">fold_change()</a></code> </p>
        </td>
        <td><p>Calculate fold change and p.value for your data.</p></td>
      </tr><tr><td>
          <p><code><a href="IgASeq-class.html">IgASeq-class</a></code> <code><a href="IgASeq-class.html">ig_seq</a></code> </p>
        </td>
        <td><p>Create a S4 object for the IgASeq data</p></td>
      </tr><tr><td>
          <p><code><a href="list_fastq.html">list_fastq()</a></code> </p>
        </td>
        <td><p>Make the list of fastq files for the rest of the proceedings.</p></td>
      </tr><tr><td>
          <p><code><a href="log_ratio.html">log_ratio()</a></code> </p>
        </td>
        <td><p>Preprocess relative abundance.</p></td>
      </tr><tr><td>
          <p><code><a href="merge_and_remove_chim.html">merge_and_remove_chim()</a></code> </p>
        </td>
        <td><p>Merge reads and remove chimera</p></td>
      </tr><tr><td>
          <p><code><a href="ml-class.html">ml-class</a></code> <code><a href="ml-class.html">ml</a></code> </p>
        </td>
        <td><p>Create a S4 object for the machine learning results.</p></td>
      </tr><tr><td>
          <p><code><a href="neg_dispersion.html">neg_dispersion()</a></code> </p>
        </td>
        <td><p>Check the dispersion of your samples and the impact of the zero_treatment.</p></td>
      </tr><tr><td>
          <p><code><a href="phylo_heatmap.html">phylo_heatmap()</a></code> </p>
        </td>
        <td><p>A function to plot heatmap nicely, using a phylogenetic distance.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_all_variance.html">plot_all_variance()</a></code> </p>
        </td>
        <td><p>Plot the overall variance of your dataset</p></td>
      </tr><tr><td>
          <p><code><a href="plot_constrained_reduction.html">plot_constrained_reduction()</a></code> </p>
        </td>
        <td><p>Make a constrained reduction on matrices and data.frame.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_corr_heatmap.html">plot_corr_heatmap()</a></code> </p>
        </td>
        <td><p>Plot a correlation heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plot_distribution.html">plot_distribution()</a></code> </p>
        </td>
        <td><p>Plot the distribution of your data.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_ig_seq.html">plot_ig_seq()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_loadings.html">plot_loadings()</a></code> </p>
        </td>
        <td><p>Title</p></td>
      </tr><tr><td>
          <p><code><a href="plot_plsda.html">plot_plsda()</a></code> </p>
        </td>
        <td><p>Plot PLS-DA using mixOmics <code>plsda</code></p></td>
      </tr><tr><td>
          <p><code><a href="plot_qual.html">plot_qual()</a></code> </p>
        </td>
        <td><p>Check the length of your reads.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_reduction.html">plot_reduction()</a></code> </p>
        </td>
        <td><p>Use multidimensional reduction on matrices and data.frame.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_volcano.html">plot_volcano()</a></code> </p>
        </td>
        <td><p>Plot volcano</p></td>
      </tr><tr><td>
          <p><code><a href="plot_xy_variance.html">plot_xy_variance()</a></code> </p>
        </td>
        <td><p>Plot the variance for two factors.</p></td>
      </tr><tr><td>
          <p><code><a href="predict_models.html">predict_models()</a></code> </p>
        </td>
        <td><p>Predict the classification based on the model training.</p></td>
      </tr><tr><td>
          <p><code><a href="qc_check.html">qc_check()</a></code> </p>
        </td>
        <td><p>A wrapper function to plot the quality of fastq.</p></td>
      </tr><tr><td>
          <p><code><a href="rng_fastq.html">rng_fastq()</a></code> </p>
        </td>
        <td><p>Randomly select fastq files</p></td>
      </tr><tr><td>
          <p><code><a href="screen_models.html">screen_models()</a></code> </p>
        </td>
        <td><p>Title Screen a model on all your variables.</p></td>
      </tr><tr><td>
          <p><code><a href="select_fastq.html">select_fastq()</a></code> </p>
        </td>
        <td><p>Select the fastq in the directory you want.</p></td>
      </tr><tr><td>
          <p><code><a href="seq_table.html">seq_table()</a></code> </p>
        </td>
        <td><p>IgASeq table</p></td>
      </tr><tr><td>
          <p><code><a href="slide_z.html">slide_z()</a></code> </p>
        </td>
        <td><p>Analyse the ratio between your IgA positive fraction and the two IgA negative
fractions.</p></td>
      </tr><tr><td>
          <p><code><a href="slide_z_modern.html">slide_z_modern()</a></code> </p>
        </td>
        <td><p>Modern version of the Z test on the IgASeq data.</p></td>
      </tr><tr><td>
          <p><code><a href="summarise_fastq.html">summarise_fastq()</a></code> </p>
        </td>
        <td><p>A function to summarise the fastq from a run.</p></td>
      </tr><tr><td>
          <p><code><a href="ternary_plot.html">ternary_plot()</a></code> </p>
        </td>
        <td><p>Plot fold change or abundance as a ternary plot</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

